diff -up evolution-3.22.6/calendar/gui/e-meeting-time-sel.c.meeting-time-selector-crash evolution-3.22.6/calendar/gui/e-meeting-time-sel.c
--- evolution-3.22.6/calendar/gui/e-meeting-time-sel.c.meeting-time-selector-crash	2016-07-11 14:31:29.000000000 +0200
+++ evolution-3.22.6/calendar/gui/e-meeting-time-sel.c	2017-04-12 14:53:43.521196869 +0200
@@ -1427,6 +1427,12 @@ e_meeting_time_selector_refresh_cb (gpoi
 {
 	EMeetingTimeSelector *mts = data;
 
+	if (!mts->model) {
+		/* Destroyed, do not do anything */
+		g_object_unref (mts);
+		return FALSE;
+	}
+
 	if (e_meeting_store_get_num_queries (mts->model) == 0) {
 		GdkCursor *cursor;
 		GdkWindow *window;
