diff -up evolution-2.28.3/calendar/gui/alarm-notify/alarm-notify.c.calendar-alarm-notifications evolution-2.28.3/calendar/gui/alarm-notify/alarm-notify.c
--- evolution-2.28.3/calendar/gui/alarm-notify/alarm-notify.c.calendar-alarm-notifications	2009-07-13 03:45:12.000000000 -0400
+++ evolution-2.28.3/calendar/gui/alarm-notify/alarm-notify.c	2012-09-07 13:51:32.267881702 -0400
@@ -372,7 +372,7 @@ alarm_notify_add_calendar (AlarmNotify *
 		const gchar *name = e_source_get_property (source, "auth-domain");
 
 		if (!name)
-			name = e_source_peek_name (source);
+			name = "Calendar";
 
 		if (!e_passwords_get_password (name, pass_key)) {
 			g_mutex_unlock (an->priv->mutex);
