diff -up evolution-2.28.3/mail/em-migrate.c.mail-migration-crash evolution-2.28.3/mail/em-migrate.c
--- evolution-2.28.3/mail/em-migrate.c.mail-migration-crash	2009-08-07 04:46:51.000000000 -0400
+++ evolution-2.28.3/mail/em-migrate.c	2011-04-27 15:48:04.524302185 -0400
@@ -232,6 +232,13 @@ struct _imap_folder_info_1_0 {
 	gchar dir_sep;
 };
 
+struct _migrate_state_info {
+	gchar *label_name;
+	double progress;
+};
+static gboolean update_states_in_main_thread (const struct 
+		_migrate_state_info *info);
+
 static GHashTable *accounts_1_0 = NULL;
 static GHashTable *accounts_name_1_0 = NULL;
 
@@ -1168,10 +1175,6 @@ em_migrate_set_folder_name (const gchar 
 	text = g_strdup_printf (_("Migrating '%s':"), folder_name);
 	gtk_label_set_text (label, text);
 	g_free (text);
-
-	gtk_progress_bar_set_fraction (progress, 0.0);
-	while (gtk_events_pending ())
-		gtk_main_iteration ();
 }
 
 static void
@@ -1523,6 +1526,13 @@ mbox_build_filename (GString *path, cons
 	}
 }
 
+static void
+migrate_state_info_free (struct _migrate_state_info *info)
+{
+	g_free (info->label_name);
+	g_slice_free (struct _migrate_state_info, info);
+}
+
 static gint
 em_migrate_folder(EMMigrateSession *session, const gchar *dirname, const gchar *full_name, CamelException *ex)
 {
@@ -1535,6 +1545,7 @@ em_migrate_folder(EMMigrateSession *sess
 	gint index, i;
 	GString *src, *dest;
 	gint res = -1;
+	struct _migrate_state_info * info;
 
 	src = g_string_new("");
 
@@ -1549,7 +1560,15 @@ em_migrate_folder(EMMigrateSession *sess
 
 	dest = g_string_new("");
 	uri = get_local_store_uri(dirname, &name, &index);
-	em_migrate_set_folder_name (full_name);
+
+	info = g_slice_new0 (struct _migrate_state_info);
+	info->label_name = g_strdup (full_name);
+
+	g_idle_add_full (
+		G_PRIORITY_LOW, (GSourceFunc)
+		update_states_in_main_thread, info,
+		(GDestroyNotify) migrate_state_info_free);
+
 	thread_list = get_local_et_expanded (dirname);
 
 	/* Manually copy local mbox files, its much faster */
@@ -2764,10 +2783,18 @@ em_update_sa_junk_setting_2_23 (void)
 }
 
 static gboolean
-update_progress_in_main_thread (double *progress)
+update_states_in_main_thread (const struct _migrate_state_info * info)
 {
-		em_migrate_set_progress (*progress);
-		return FALSE;
+	g_return_val_if_fail (info != NULL, FALSE);
+	g_return_val_if_fail (info->label_name != NULL, FALSE);
+	em_migrate_set_progress (info->progress);
+	em_migrate_set_folder_name (info->label_name);
+
+	/* XXX Why? */
+	while (gtk_events_pending ())
+		gtk_main_iteration ();
+
+	return FALSE;
 }
 
 static void
@@ -2776,17 +2803,18 @@ migrate_folders(CamelStore *store, gbool
 	CamelFolder *folder;
 
 	while (fi) {
-		double progress;
-		gchar *tmp;
+		struct _migrate_state_info *info;
 
 		*nth_folder = *nth_folder + 1;
 
-		tmp = g_strdup_printf ("%s/%s", acc, fi->full_name);
-		em_migrate_set_folder_name (tmp);
-		g_free (tmp);
-
-		progress = (double) (*nth_folder) / total_folders;
-		g_idle_add ((GSourceFunc) update_progress_in_main_thread, &progress);
+		info = g_slice_new0 (struct _migrate_state_info);
+		info->label_name = g_strdup_printf ("%s/%s", acc, fi->full_name);
+		info->progress = (double) (*nth_folder) / total_folders;
+
+		g_idle_add_full (
+			G_PRIORITY_LOW, (GSourceFunc)
+			update_states_in_main_thread, info,
+			(GDestroyNotify) migrate_state_info_free);
 
 		if (is_local)
 				folder = camel_store_get_folder (store, fi->full_name, CAMEL_STORE_IS_MIGRATING, ex);
