commit f8eff85f768e336913bfb9d3027ed33105c573fd
Author: Matthew Barnes <mbarnes@redhat.com>
Date:   Fri Jan 21 09:08:24 2011 -0500

    evolution-2.28.3-crash-on-large-message.patch

diff --git a/mail/em-format-html-display.c b/mail/em-format-html-display.c
index 003eb37..b8f0c55 100644
--- a/mail/em-format-html-display.c
+++ b/mail/em-format-html-display.c
@@ -1698,7 +1698,7 @@ efhd_format_optional(EMFormat *emf, CamelStream *fstream, CamelMimePart *part, C
 	info->snoop_mime_type = "text/plain";
 	info->attachment = e_attachment_new ();
 	e_attachment_set_mime_part (info->attachment, info->puri.part);
-	info->mstream = (CamelStreamMem *)mstream;
+	info->mstream = (CamelStreamMem *) mstream; camel_object_ref (mstream);
 	if (emf->valid) {
 		info->sign = emf->valid->sign.status;
 		info->encrypt = emf->valid->encrypt.status;
