diff -up evolution-3.12.11/shell/e-shell.c.remove-skip-lookup-button evolution-3.12.11/shell/e-shell.c
--- evolution-3.12.11/shell/e-shell.c.remove-skip-lookup-button	2014-09-15 13:08:34.000000000 +0200
+++ evolution-3.12.11/shell/e-shell.c	2015-05-04 18:35:32.656094474 +0200
@@ -823,14 +823,6 @@ shell_window_added (GtkApplication *appl
 		(gintptr) window);
 	gtk_window_set_role (window, role);
 	g_free (role);
-
-#if GTK_CHECK_VERSION(3,11,5)
-	/* Forbid header bars in stock GTK+ dialogs.
-	 * They look very out of place in Evolution. */
-	g_object_set (
-		gtk_widget_get_settings (GTK_WIDGET (window)),
-		"gtk-dialogs-use-header", FALSE, NULL);
-#endif
 }
 
 static gboolean
@@ -874,6 +866,14 @@ shell_initable_init (GInitable *initable
 	g_object_unref (proxy_source);
 	g_object_unref (registry);
 
+#if GTK_CHECK_VERSION(3,11,5)
+	/* Forbid header bars in stock GTK+ dialogs.
+	 * They look very out of place in Evolution. */
+	g_object_set (
+		gtk_settings_get_default (),
+		"gtk-dialogs-use-header", FALSE, NULL);
+#endif
+
 	return TRUE;
 }
 
