diff -up evolution-2.32.3/m4/evo_ldap_check.m4.ldap-with-nss evolution-2.32.3/m4/evo_ldap_check.m4
--- evolution-2.32.3/m4/evo_ldap_check.m4.ldap-with-nss	2011-03-29 20:38:48.000000000 +0200
+++ evolution-2.32.3/m4/evo_ldap_check.m4	2013-05-29 11:35:03.111613808 +0200
@@ -55,7 +55,8 @@ AC_DEFUN([EVO_LDAP_CHECK],[
 			AC_MSG_ERROR([evolution requires OpenLDAP version >= 2])
 		fi
 
-		AC_CHECK_LIB(resolv, res_query, [LDAP_LIBS="-lresolv"])
+		LDAP_LIBS="-lssl3 -lsmime3 -lnss3 -lnssutil3 -lplds4 -lplc4 -lnspr4"
+		AC_CHECK_LIB(resolv, res_query, [LDAP_LIBS="$LDAP_LIBS -lresolv"])
 		AC_CHECK_LIB(socket, bind, [LDAP_LIBS="$LDAP_LIBS -lsocket"])
 		AC_CHECK_LIB(nsl, gethostbyaddr, [LDAP_LIBS="$LDAP_LIBS -lnsl"])
 		AC_CHECK_LIB(lber, ber_get_tag, [
