diff --git a/calendar/gui/print.c b/calendar/gui/print.c
index 57a1dc1..095362b 100644
--- a/calendar/gui/print.c
+++ b/calendar/gui/print.c
@@ -49,6 +49,7 @@
 #include "e-day-view-layout.h"
 #include "e-week-view.h"
 #include "e-week-view-layout.h"
+#include "e-task-table.h"
 #include "gnome-cal.h"
 #include "print.h"
 
@@ -60,6 +61,8 @@ typedef struct PrintCalItem PrintCalItem;
 struct PrintCompItem {
 	ECal *client;
 	ECalComponent *comp;
+	icaltimezone *zone;
+	gboolean use_24_hour_format;
 };
 
 struct PrintCalItem {
@@ -87,6 +90,26 @@ evo_calendar_print_renderer_get_width (GtkPrintContext *context,
 	return pango_units_to_double (layout_width);
 }
 
+static gdouble
+evo_calendar_print_renderer_get_height (GtkPrintContext *context,
+					PangoFontDescription *font,
+					const gchar *text)
+{
+	PangoLayout *layout;
+	gint layout_width, layout_height;
+
+	layout = gtk_print_context_create_pango_layout (context);
+
+	pango_layout_set_font_description (layout, font);
+	pango_layout_set_text (layout, text, -1);
+	pango_layout_set_indent (layout, 0);
+	pango_layout_get_size (layout, &layout_width, &layout_height);
+
+	g_object_unref (layout);
+
+	return pango_units_to_double (layout_height);
+}
+
 static double
 get_font_size (PangoFontDescription *font)
 {
@@ -123,7 +146,7 @@ get_font_size (PangoFontDescription *font)
 
 /* The width of the small calendar months, the space from the right edge of
    the header rectangle, and the space between the months. */
-#define SMALL_MONTH_WIDTH	100
+#define MIN_SMALL_MONTH_WIDTH	100
 #define SMALL_MONTH_PAD		5
 #define SMALL_MONTH_SPACING	20
 
@@ -177,6 +200,7 @@ struct pdinfo
 	gint mins_per_row;
 	guint8 cols_per_row[DAY_VIEW_ROWS];
 	gboolean use_24_hour_format;
+	icaltimezone *zone;
 };
 
 struct psinfo
@@ -196,31 +220,31 @@ struct psinfo
 	gboolean use_24_hour_format;
 	gdouble row_height;
 	gdouble header_row_height;
+	icaltimezone *zone;
 };
 
 /* Convenience function to help the transition to timezone functions.
    It converts a time_t to a struct tm. */
-static struct tm*
-convert_timet_to_struct_tm (time_t time, icaltimezone *zone)
+static void
+convert_timet_to_struct_tm (time_t time,
+                            icaltimezone *zone,
+                            struct tm *tm)
 {
-	static struct tm my_tm;
 	struct icaltimetype tt;
 
 	/* Convert it to an icaltimetype. */
 	tt = icaltime_from_timet_with_zone (time, FALSE, zone);
 
 	/* Fill in the struct tm. */
-	my_tm.tm_year = tt.year - 1900;
-	my_tm.tm_mon = tt.month - 1;
-	my_tm.tm_mday = tt.day;
-	my_tm.tm_hour = tt.hour;
-	my_tm.tm_min = tt.minute;
-	my_tm.tm_sec = tt.second;
-	my_tm.tm_isdst = tt.is_daylight;
-
-	my_tm.tm_wday = time_day_of_week (tt.day, tt.month - 1, tt.year);
-
-	return &my_tm;
+	tm->tm_year = tt.year - 1900;
+	tm->tm_mon = tt.month - 1;
+	tm->tm_mday = tt.day;
+	tm->tm_hour = tt.hour;
+	tm->tm_min = tt.minute;
+	tm->tm_sec = tt.second;
+	tm->tm_isdst = tt.is_daylight;
+
+	tm->tm_wday = time_day_of_week (tt.day, tt.month - 1, tt.year);
 }
 
 /* Fills the 42-element days array with the day numbers for the specified
@@ -228,7 +252,12 @@ convert_timet_to_struct_tm (time_t time, icaltimezone *zone)
  * The starting and ending indexes of the days are returned in the start
  * and end arguments. */
 static void
-build_month (gint month, gint year, gint *days, gint *start, gint *end)
+build_month (ECalModel *model,
+             gint month,
+             gint year,
+             gint *days,
+             gint *start,
+             gint *end)
 {
 	gint i;
 	gint d_month, d_week, week_start_day;
@@ -256,7 +285,7 @@ build_month (gint month, gint year, gint *days, gint *start, gint *end)
 
 	/* Get the configuration setting specifying which weekday we put on
 	   the left column, 0=Sun to 6=Sat. */
-	week_start_day = calendar_config_get_week_start_day ();
+	week_start_day = e_cal_model_get_week_start_day (model);
 
 	/* Figure out which square we want to put the 1 in. */
 	d_week = (d_week + 7 - week_start_day) % 7;
@@ -438,7 +467,7 @@ print_text (GtkPrintContext *context, PangoFontDescription *desc,
 	cairo_move_to (cr, x1, y1);
 	pango_cairo_show_layout (cr, layout);
 
-	cairo_stroke(cr);
+	cairo_stroke (cr);
 
 	cairo_restore (cr);
 
@@ -479,7 +508,6 @@ print_text_size_bold (GtkPrintContext *context, const gchar *text,
 	return w;
 }
 
-#if 0  /* KILL-BONOBO */
 static void
 titled_box (GtkPrintContext *context, const gchar *text,
             PangoFontDescription *font, PangoAlignment alignment,
@@ -488,21 +516,27 @@ titled_box (GtkPrintContext *context, const gchar *text,
 {
 	gdouble size;
 
-	size = get_font_size (font);
-	print_border (context, *x1, *x2, *y1, *y1 + size * 1.4, linewidth, 0.9);
-	print_border (context, *x1, *x2, *y1 + size * 1.4, *y2, linewidth, -1.0);
+	size = evo_calendar_print_renderer_get_height (context, font, text);
+	print_border (context, *x1, *x2, *y1, *y1 + size  + 2, linewidth, 0.9);
+	print_border (context, *x1, *x2, *y1 + size + 2, *y2, linewidth, -1.0);
 	*x1 += 2;
 	*x2 -= 2;
 	*y2 += 2;
-	print_text (context, font, text, alignment, *x1, *x2, *y1 + 1.0, *y1 + size * 1.4);
-	*y1 += size * 1.4;
+	print_text (context, font, text, alignment, *x1, *x2, *y1 + 1.0, *y1 + size);
+	*y1 += size + 2;
 }
-#endif /* KILL-BONOBO */
 
 static gboolean
 get_show_week_numbers (void)
 {
-	return e_shell_settings_get_boolean (e_shell_get_shell_settings (e_shell_get_default ()), "cal-show-week-numbers");
+	EShell *shell;
+	EShellSettings *shell_settings;
+
+	shell = e_shell_get_default ();
+	shell_settings = e_shell_get_shell_settings (shell);
+
+	return e_shell_settings_get_boolean (
+		shell_settings, "cal-show-week-numbers");
 }
 
 enum datefmt {
@@ -526,13 +560,12 @@ static const gchar *days[] = {
   format the date 'nicely' and consistently for various headers
 */
 static gchar *
-format_date(time_t time, gint flags, gchar *buffer, gint bufflen)
+format_date (struct tm *tm,
+             gint flags,
+             gchar *buffer,
+             gint bufflen)
 {
-	icaltimezone *zone = calendar_config_get_icaltimezone ();
 	gchar fmt[64];
-	struct tm tm;
-
-	tm = *convert_timet_to_struct_tm (time, zone);
 
 	fmt[0] = 0;
 	if (flags & DATE_DAYNAME) {
@@ -541,7 +574,7 @@ format_date(time_t time, gint flags, gchar *buffer, gint bufflen)
 	if (flags & DATE_DAY) {
 		if (flags & DATE_DAYNAME)
 			strcat(fmt, " ");
-		strcat(fmt, gettext(days[tm.tm_mday-1]));
+		strcat (fmt, gettext (days[tm->tm_mday-1]));
 	}
 	if (flags & DATE_MONTH) {
 		if (flags & (DATE_DAY|DATE_DAYNAME))
@@ -555,7 +588,7 @@ format_date(time_t time, gint flags, gchar *buffer, gint bufflen)
 			strcat(fmt, " ");
 		strcat(fmt, "%Y");
 	}
-	e_utf8_strftime(buffer, bufflen, fmt, &tm);
+	e_utf8_strftime (buffer, bufflen, fmt, tm);
 	buffer[bufflen - 1] = '\0';
 
 	return buffer;
@@ -575,6 +608,39 @@ instance_cb (ECalComponent *comp,
 	return FALSE;
 }
 
+/* Translators: These are workday abbreviations, e.g. Su=Sunday and Th=thursday */
+const gchar *daynames[] =
+	{ N_("Su"), N_("Mo"), N_("Tu"), N_("We"),
+	  N_("Th"), N_("Fr"), N_("Sa") };
+
+static gdouble
+calc_small_month_width (GtkPrintContext *context, gdouble for_height)
+{
+
+	PangoFontDescription *font_bold;
+	gdouble res = 0.0;
+	gint ii;
+
+	font_bold = get_font_for_size (for_height / 7.4, PANGO_WEIGHT_BOLD);
+	res = MAX (evo_calendar_print_renderer_get_width (
+		context, font_bold, "23"), res);
+	for (ii = 0; ii < 7; ii++) {
+		res = MAX (evo_calendar_print_renderer_get_width (
+			context, font_bold, _(daynames[ii])), res);
+	}
+
+	pango_font_description_free (font_bold);
+
+	/* res is max cell width, so multiply it with column
+	 * count plus some space between columns. */
+	res = (res + 1.0) * (7 + (get_show_week_numbers () ? 1 : 0)) - 1.0;
+
+	if (res < MIN_SMALL_MONTH_WIDTH)
+		res = MIN_SMALL_MONTH_WIDTH;
+
+	return res;
+}
+
 /*
   print out the month small, embolden any days with events.
 */
@@ -584,8 +650,9 @@ print_month_small (GtkPrintContext *context, GnomeCalendar *gcal, time_t month,
 		   gint titleflags, time_t greystart, time_t greyend,
 		   gint bordertitle)
 {
-	icaltimezone *zone = calendar_config_get_icaltimezone ();
+	icaltimezone *zone;
 	PangoFontDescription *font, *font_bold, *font_normal;
+	ECalModel *model;
 	time_t now, next;
 	gint x, y;
 	gint days[42];
@@ -596,17 +663,16 @@ print_month_small (GtkPrintContext *context, GnomeCalendar *gcal, time_t month,
 	gdouble header_size, col_width, row_height, text_xpad, w;
 	gdouble cell_top, cell_bottom, cell_left, cell_right, text_right;
 	gboolean week_numbers;
-
-	/* Translators: These are workday abbreviations, e.g. Su=Sunday and Th=thursday */
-	const gchar *daynames[] =
-		{ N_("Su"), N_("Mo"), N_("Tu"), N_("We"),
-		  N_("Th"), N_("Fr"), N_("Sa") };
 	cairo_t *cr;
 
+	model = gnome_calendar_get_model (gcal);
+	zone = e_cal_model_get_timezone (model);
+
 	week_numbers = get_show_week_numbers ();
 
 	/* Print the title, e.g. 'June 2001', in the top 16% of the area. */
-	format_date (month, titleflags, buf, 100);
+	convert_timet_to_struct_tm (month, zone, &tm);
+	format_date (&tm, titleflags, buf, 100);
 
 	header_size = ABS (y2 - y1) * 0.16;
 
@@ -629,8 +695,8 @@ print_month_small (GtkPrintContext *context, GnomeCalendar *gcal, time_t month,
 	font_size = row_height;
 
 	/* get month days */
-	tm = *convert_timet_to_struct_tm (month, zone);
-	build_month (tm.tm_mon, tm.tm_year + 1900, days, NULL, NULL);
+	convert_timet_to_struct_tm (month, zone, &tm);
+	build_month (model, tm.tm_mon, tm.tm_year + 1900, days, NULL, NULL);
 
 	font_normal = get_font_for_size (font_size, PANGO_WEIGHT_NORMAL);
 	font_bold = get_font_for_size (font_size, PANGO_WEIGHT_BOLD);
@@ -645,13 +711,14 @@ print_month_small (GtkPrintContext *context, GnomeCalendar *gcal, time_t month,
 	cairo_set_source_rgb (cr, 0, 0, 0);
 
 	/* Print the abbreviated day names across the top in bold. */
-	week_start_day = calendar_config_get_week_start_day ();
+	week_start_day = e_cal_model_get_week_start_day (model);
 	weekday = week_start_day;
 	for (x = 0; x < 7; x++) {
 		print_text (
 			context, font_bold,
 			_(daynames[weekday]), PANGO_ALIGN_RIGHT,
-			x1 + (x + (week_numbers ? 1 : 0)) * col_width, x1 + (x + 1 + (week_numbers ? 1 : 0)) * col_width,
+			x1 + (x + (week_numbers ? 1 : 0)) * col_width,
+			x1 + (x + 1 + (week_numbers ? 1 : 0)) * col_width,
 			y1, y1 + row_height * 1.4);
 		weekday = (weekday + 1) % 7;
 	}
@@ -683,10 +750,14 @@ print_month_small (GtkPrintContext *context, GnomeCalendar *gcal, time_t month,
 			if (day != 0) {
 				time_t week_begin = time_week_begin_with_zone (now, week_start_day, zone);
 
-				tm = *convert_timet_to_struct_tm (week_begin, zone);
+				convert_timet_to_struct_tm (week_begin, zone, &tm);
 
 				/* month in e_calendar_item_get_week_number is also zero-based */
-				sprintf (buf, "%d", e_calendar_item_get_week_number (NULL, tm.tm_mday, tm.tm_mon, tm.tm_year + 1900));
+				sprintf (
+					buf, "%d",
+					e_calendar_item_get_week_number (
+					NULL, tm.tm_mday, tm.tm_mon,
+					tm.tm_year + 1900));
 
 				print_text (context, font_normal, buf, PANGO_ALIGN_RIGHT,
 					    cell_left, text_right,
@@ -799,6 +870,7 @@ print_day_background (GtkPrintContext *context, GnomeCalendar *gcal,
 		      time_t whence, struct pdinfo *pdi,
 		      double left, double right, double top, double bottom)
 {
+	ECalModel *model;
 	PangoFontDescription *font_hour, *font_minute;
 	gdouble yinc, y;
 	gdouble width = DAY_VIEW_TIME_COLUMN_WIDTH;
@@ -807,9 +879,12 @@ print_day_background (GtkPrintContext *context, GnomeCalendar *gcal,
 	const gchar *minute;
 	gboolean use_24_hour;
 	gint i, hour, row;
-	gdouble hour_minute_x;
+	gdouble hour_minute_x, hour_minute_width;
 	cairo_t *cr;
 
+	model = gnome_calendar_get_model (gcal);
+	use_24_hour = calendar_config_get_24_hour_format ();
+
 	/* Fill the time column in light-gray. */
 	print_border (context, left, left + width, top, bottom, -1.0, 0.9);
 
@@ -834,19 +909,24 @@ print_day_background (GtkPrintContext *context, GnomeCalendar *gcal,
 
         /* Get the 2 fonts we need. */
 	font_size = yinc * 0.6;
-	max_font_size = width * 0.5;
+	max_font_size = width * 0.45;
 	hour_font_size = MIN (font_size, max_font_size);
 	font_hour = get_font_for_size (hour_font_size, PANGO_WEIGHT_BOLD);
 
 	font_size = yinc * 0.33;
-	max_font_size = width * 0.25;
+	max_font_size = width * 0.2;
 	minute_font_size = MIN (font_size, max_font_size);
 	font_minute = get_font_for_size (minute_font_size, PANGO_WEIGHT_BOLD);
-
-	use_24_hour = calendar_config_get_24_hour_format ();
+	hour_minute_width = evo_calendar_print_renderer_get_width (
+		context, font_minute, use_24_hour ? "00" : _("am"));
+	if (!use_24_hour)
+		hour_minute_width =
+			MAX (hour_minute_width,
+			evo_calendar_print_renderer_get_width (
+			context, font_minute, _("pm")));
 
 	row = 0;
-	hour_minute_x = left + width * 0.58;
+	hour_minute_x = left + width - hour_minute_width - 3;
 	for (i = pdi->start_hour; i < pdi->end_hour; i++) {
 		y = top + yinc * (row + 1);
 		cr = gtk_print_context_get_cairo_context (context);
@@ -901,13 +981,13 @@ static gint
 print_day_add_event (ECalModelComponent *comp_data,
 		     time_t	    start,
 		     time_t	    end,
+		     icaltimezone *zone,
 		     gint	    days_shown,
 		     time_t	   *day_starts,
 		     GArray	   *long_events,
 		     GArray	  **events)
 
 {
-	icaltimezone *zone = calendar_config_get_icaltimezone ();
 	EDayViewEvent event;
 	gint day, offset;
 	struct icaltimetype start_tt, end_tt;
@@ -978,7 +1058,7 @@ print_day_details_cb (ECalComponent *comp, time_t istart, time_t iend,
 	struct pdinfo *pdi = (struct pdinfo *) mdata->cb_data;
 
 	print_day_add_event (mdata->comp_data, istart, iend,
-			     pdi->days_shown, pdi->day_starts,
+			     pdi->zone, pdi->days_shown, pdi->day_starts,
 			     pdi->long_events, pdi->events);
 
 	return TRUE;
@@ -993,7 +1073,7 @@ free_event_array (GArray *array)
 	for (event_num = 0; event_num < array->len; event_num++) {
 		event = &g_array_index (array, EDayViewEvent, event_num);
 		if (event->canvas_item)
-			gtk_object_destroy (GTK_OBJECT (event->canvas_item));
+			g_object_run_dispose (G_OBJECT (event->canvas_item));
 	}
 
 	g_array_set_size (array, 0);
@@ -1291,11 +1371,12 @@ static void
 print_day_details (GtkPrintContext *context, GnomeCalendar *gcal, time_t whence,
 		   double left, double right, double top, double bottom)
 {
-	icaltimezone *zone = calendar_config_get_icaltimezone ();
+	ECalModel *model;
+	icaltimezone *zone;
 	EDayViewEvent *event;
 	PangoFontDescription *font;
 	time_t start, end;
-	struct pdinfo pdi;
+	struct pdinfo pdi = { 0 };
 	gint rows_in_top_display, i;
 	gdouble font_size, max_font_size;
 	cairo_t *cr;
@@ -1303,7 +1384,8 @@ print_day_details (GtkPrintContext *context, GnomeCalendar *gcal, time_t whence,
 #define LONG_DAY_EVENTS_TOP_SPACING 4
 #define LONG_DAY_EVENTS_BOTTOM_SPACING 2
 
-	ECalModel *model = gnome_calendar_get_model (gcal);
+	model = gnome_calendar_get_model (gcal);
+	zone = e_cal_model_get_timezone (model);
 
 	start = time_day_begin_with_zone (whence, zone);
 	end = time_day_end_with_zone (start, zone);
@@ -1322,6 +1404,7 @@ print_day_details (GtkPrintContext *context, GnomeCalendar *gcal, time_t whence,
 	pdi.start_minute_offset = pdi.start_hour * 60;
 	pdi.end_minute_offset = pdi.end_hour * 60;
 	pdi.use_24_hour_format = calendar_config_get_24_hour_format ();
+	pdi.zone = e_cal_model_get_timezone (model);
 
 	/* Get the events from the server. */
 	e_cal_model_generate_instances (model, start, end, print_day_details_cb, &pdi);
@@ -1442,7 +1525,7 @@ print_day_details (GtkPrintContext *context, GnomeCalendar *gcal, time_t whence,
 		max_font_size = ((top - bottom) / pdi.rows) - 4;
 	else
 		max_font_size = ((bottom - top ) / pdi.rows) - 4;
-	font_size = MIN(DAY_NORMAL_FONT_SIZE, max_font_size);
+	font_size = MIN (DAY_NORMAL_FONT_SIZE, max_font_size);
 	font = get_font_for_size (font_size, PANGO_WEIGHT_NORMAL);
 
 	for (i = 0; i < pdi.events[0]->len; i++) {
@@ -1730,7 +1813,6 @@ print_week_view_background (GtkPrintContext *context,
 			    double left, double top,
 			    double cell_width, double cell_height)
 {
-	icaltimezone *zone = calendar_config_get_icaltimezone ();
 	struct tm tm;
 	gint day, day_x, day_y, day_h;
 	gdouble x1, x2, y1, y2, font_size, fillcolor;
@@ -1751,7 +1833,7 @@ print_week_view_background (GtkPrintContext *context,
 		y1 = top + day_y * cell_height;
 		y2 = y1 + day_h * cell_height;
 
-		tm = *convert_timet_to_struct_tm (psi->day_starts[day], zone);
+		convert_timet_to_struct_tm (psi->day_starts[day], psi->zone, &tm);
 
 		/* In the month view we draw a grey background for the end
 		   of the previous month and the start of the following. */
@@ -1801,7 +1883,6 @@ print_week_summary_cb (ECalComponent *comp,
 		       gpointer	  data)
 
 {
-	icaltimezone *zone = calendar_config_get_icaltimezone ();
 	EWeekViewEvent event;
 	struct icaltimetype start_tt, end_tt;
 	ECalModelGenerateInstancesData *mdata = (ECalModelGenerateInstancesData *) data;
@@ -1819,8 +1900,8 @@ print_week_summary_cb (ECalComponent *comp,
 	g_return_val_if_fail (start < psi->day_starts[psi->days_shown], TRUE);
 	g_return_val_if_fail (end > psi->day_starts[0], TRUE);
 
-	start_tt = icaltime_from_timet_with_zone (start, FALSE, zone);
-	end_tt = icaltime_from_timet_with_zone (end, FALSE, zone);
+	start_tt = icaltime_from_timet_with_zone (start, FALSE, psi->zone);
+	end_tt = icaltime_from_timet_with_zone (end, FALSE, psi->zone);
 
 	event.comp_data = g_object_ref (mdata->comp_data);
 
@@ -1845,21 +1926,25 @@ print_week_summary (GtkPrintContext *context, GnomeCalendar *gcal,
 		    gint month, double font_size,
 		    double left, double right, double top, double bottom)
 {
-	icaltimezone *zone = calendar_config_get_icaltimezone ();
+	icaltimezone *zone;
 	EWeekViewEvent *event;
-	struct psinfo psi;
+	struct psinfo psi = { 0 };
 	time_t day_start;
 	gint rows_per_day[E_WEEK_VIEW_MAX_WEEKS * 7], day, event_num;
 	GArray *spans;
 	PangoFontDescription *font;
 	gdouble cell_width, cell_height;
-	ECalModel *model = gnome_calendar_get_model (gcal);
+	ECalModel *model;
+
+	model = gnome_calendar_get_model (gcal);
+	zone = e_cal_model_get_timezone (model);
 
 	psi.days_shown = weeks_shown * 7;
 	psi.events = g_array_new (FALSE, FALSE, sizeof (EWeekViewEvent));
 	psi.multi_week_view = multi_week_view;
 	psi.weeks_shown = weeks_shown;
 	psi.month = month;
+	psi.zone = zone;
 
 	/* Get a few config settings. */
 	if (multi_week_view)
@@ -1869,7 +1954,7 @@ print_week_summary (GtkPrintContext *context, GnomeCalendar *gcal,
 	psi.use_24_hour_format = calendar_config_get_24_hour_format ();
 
 	/* We convert this from (0 = Sun, 6 = Sat) to (0 = Mon, 6 = Sun). */
-	psi.display_start_weekday = calendar_config_get_week_start_day ();
+	psi.display_start_weekday = e_cal_model_get_week_start_day (model);
 	psi.display_start_weekday = (psi.display_start_weekday + 6) % 7;
 
 	/* If weekends are compressed then we can't start on a Sunday. */
@@ -1939,67 +2024,29 @@ print_week_summary (GtkPrintContext *context, GnomeCalendar *gcal,
 	g_array_free (spans, TRUE);
 }
 
-/* XXX Evolution doesn't have a "year" view. */
-#if 0
-static void
-print_year_summary (GtkPrintContext *context, GnomeCalendar *gcal, time_t whence,
-		    double left, double right, double top, double bottom,
-		    gint morerows)
-{
-	icaltimezone *zone = calendar_config_get_icaltimezone ();
-	gdouble row_height, col_width, l, r, t, b;
-	time_t now;
-	gint col, row, rows, cols;
-
-	l = left;
-	t = top;
-
-	/* If morerows is set we do 4 rows and 3 columns instead of 3 rows and
-	   4 columns. This is useful if we switch paper orientation. */
-	if (morerows) {
-		rows = 4;
-		cols = 3;
-	} else {
-		rows = 3;
-		cols = 4;
-	}
-
-	row_height = (top - bottom) / rows;
-	col_width = (right - left) / cols;
-	r = l + col_width;
-	b = top - row_height;
-	now = time_year_begin_with_zone (whence, zone);
-
-	for (row = 0; row < rows; row++) {
-		t = top - row_height * row;
-		b = t - row_height;
-		for (col = 0; col < cols; col++) {
-			l = left + col_width * col;
-			r = l + col_width;
-			print_month_small (context, gcal, now,
-					   l + 8, t - 8, r - 8 - (get_show_week_numbers () ? 1 : 0), b + 8,
-					   DATE_MONTH, 0, 0, TRUE);
-			now = time_add_month_with_zone (now, 1, zone);
-		}
-	}
-}
-#endif
-
 static void
-print_month_summary (GtkPrintContext *context, GnomeCalendar *gcal, time_t whence,
-		     double left, double right, double top, double bottom)
+print_month_summary (GtkPrintContext *context,
+                     GnomeCalendar *gcal,
+                     time_t whence,
+                     gdouble left,
+                     gdouble right,
+                     gdouble top,
+                     gdouble bottom)
 {
-	icaltimezone *zone = calendar_config_get_icaltimezone ();
+	icaltimezone *zone;
 	time_t date;
 	struct tm tm;
 	struct icaltimetype tt;
 	gchar buffer[100];
+	ECalModel *model;
 	PangoFontDescription *font;
 	gboolean compress_weekend;
 	gint columns, col, weekday, month, weeks;
 	gdouble font_size, cell_width, x1, x2, y1, y2;
 
-	weekday = calendar_config_get_week_start_day ();
+	model = gnome_calendar_get_model (gcal);
+	zone = e_cal_model_get_timezone (model);
+	weekday = e_cal_model_get_week_start_day (model);
 	compress_weekend = calendar_config_get_compress_weekend ();
 
 	date = 0;
@@ -2041,7 +2088,7 @@ print_month_summary (GtkPrintContext *context, GnomeCalendar *gcal, time_t whenc
 	   to be able to step through the week without worrying about
 	   overflows making strftime choke, so we move near to the start of
 	   the month. */
-	tm = *convert_timet_to_struct_tm (date, zone);
+	convert_timet_to_struct_tm (date, zone, &tm);
 	tm.tm_mday = (tm.tm_mday % 7) + 7;
 
 	font = get_font_for_size (MONTH_NORMAL_FONT_SIZE, PANGO_WEIGHT_BOLD);
@@ -2086,11 +2133,10 @@ print_todo_details (GtkPrintContext *context, GnomeCalendar *gcal,
 		    time_t start, time_t end,
 		    double left, double right, double top, double bottom)
 {
-#if 0  /* KILL-BONOBO */
 	PangoFontDescription *font_summary;
 	gdouble y, yend, x, xend;
 	struct icaltimetype *tt;
-	ECalendarTable *task_pad;
+	GtkWidget *task_table;
 	ETable *table;
 	ECalModel *model;
 	gint rows, row;
@@ -2098,9 +2144,10 @@ print_todo_details (GtkPrintContext *context, GnomeCalendar *gcal,
 
 	/* We get the tasks directly from the TaskPad ETable. This means we
 	   get them filtered & sorted for free. */
-	task_pad = gnome_calendar_get_task_pad (gcal);
-	table = e_calendar_table_get_table (task_pad);
-	model = e_calendar_table_get_model (task_pad);
+	task_table = gnome_calendar_get_task_table (gcal);
+	table = E_TABLE (task_table);
+	g_return_if_fail (table != NULL);
+	model = e_task_table_get_model (E_TASK_TABLE (task_table));
 
 	font_summary = get_font_for_size (12, PANGO_WEIGHT_NORMAL);
 
@@ -2176,80 +2223,531 @@ print_todo_details (GtkPrintContext *context, GnomeCalendar *gcal,
 	}
 
 	pango_font_description_free (font_summary);
-#endif
 }
 
 static void
 print_day_view (GtkPrintContext *context, GnomeCalendar *gcal, time_t date)
 {
+	ECalModel *model;
 	GtkPageSetup *setup;
-	icaltimezone *zone = calendar_config_get_icaltimezone ();
+	icaltimezone *zone;
 	gint i, days = 1;
-	gdouble todo, l, week_numbers_inc;
+	gdouble todo, l, week_numbers_inc, small_month_width;
 	gchar buf[100];
 	gdouble width, height;
+	struct tm tm;
+
+	model = gnome_calendar_get_model (gcal);
+	zone = e_cal_model_get_timezone (model);
 
 	setup = gtk_print_context_get_page_setup (context);
 
 	width = gtk_page_setup_get_page_width (setup, GTK_UNIT_POINTS);
 	height = gtk_page_setup_get_page_height (setup, GTK_UNIT_POINTS);
-	week_numbers_inc = get_show_week_numbers () ? SMALL_MONTH_WIDTH / 7.0 : 0;
+	small_month_width = calc_small_month_width (context, HEADER_HEIGHT);
+	week_numbers_inc = get_show_week_numbers () ? small_month_width / 7.0 : 0;
 
 	for (i = 0; i < days; i++) {
 		todo = width * 0.75;
 
 		/* Print the main view with all the events in. */
 		print_day_details (context, gcal, date,
-				   0.0, todo - 2.0, HEADER_HEIGHT,
+				   0.0, todo - 2.0, HEADER_HEIGHT + 4,
 				   height);
 
 		 /* Print the TaskPad down the right. */
 		print_todo_details (context, gcal, 0, INT_MAX,
-				    todo, width, HEADER_HEIGHT,
+				    todo, width, HEADER_HEIGHT + 4,
 				    height);
 
 		/* Print the filled border around the header. */
 		print_border (context, 0.0, width,
-			      0.0, HEADER_HEIGHT + 3.5, 1.0, 0.9);
+			      0.0, HEADER_HEIGHT + 4, 1.0, 0.9);
 
 		/* Print the 2 mini calendar-months. */
-		l = width - SMALL_MONTH_PAD - (SMALL_MONTH_WIDTH + week_numbers_inc) * 2  - SMALL_MONTH_SPACING;
+		l = width - SMALL_MONTH_PAD -
+			(small_month_width + week_numbers_inc) * 2 -
+			SMALL_MONTH_SPACING;
 
-		 print_month_small (context, gcal, date,
-				   l, 4, l + SMALL_MONTH_WIDTH + week_numbers_inc, HEADER_HEIGHT + 4,
+		print_month_small (context, gcal, date,
+				   l, 2, l + small_month_width + week_numbers_inc, HEADER_HEIGHT + 2,
 				   DATE_MONTH | DATE_YEAR, date, date, FALSE);
 
-		l += SMALL_MONTH_SPACING + SMALL_MONTH_WIDTH + week_numbers_inc;
+		l += SMALL_MONTH_SPACING + small_month_width + week_numbers_inc;
 		print_month_small (context, gcal,
 				   time_add_month_with_zone (date, 1, zone),
-				   l, 4, l + SMALL_MONTH_WIDTH + week_numbers_inc, HEADER_HEIGHT + 4,
+				   l, 2, l + small_month_width + week_numbers_inc, HEADER_HEIGHT + 2,
 				   DATE_MONTH | DATE_YEAR, 0, 0, FALSE);
 
 		/* Print the date, e.g. '8th May, 2001'. */
-		format_date (date, DATE_DAY | DATE_MONTH | DATE_YEAR,
+		convert_timet_to_struct_tm (date, zone, &tm);
+		format_date (&tm, DATE_DAY | DATE_MONTH | DATE_YEAR,
 			     buf, 100);
 
 		print_text_size_bold (context, buf, PANGO_ALIGN_LEFT,
 				      4, todo, 4,
-                                      4 + 24);
+				      4 + 24);
 
 		/* Print the day, e.g. 'Tuesday'. */
-		format_date (date, DATE_DAYNAME, buf, 100);
+		format_date (&tm, DATE_DAYNAME, buf, 100);
 
 		print_text_size_bold (context, buf, PANGO_ALIGN_LEFT,
-				 4, todo, 32,
-                                 32 + 18);
+				      4, todo,
+				      HEADER_HEIGHT + 9,
+				      HEADER_HEIGHT + 9 + 18);
 
 		date = time_add_day_with_zone (date, 1, zone);
 	 }
 }
 
 static void
+print_work_week_background (GtkPrintContext *context, GnomeCalendar *gcal,
+			    time_t whence, struct pdinfo *pdi, double left,
+			    double right, double top, double bottom)
+{
+	ECalModel *model;
+	PangoFontDescription *font_hour, *font_minute;
+	gdouble yinc, y;
+	gdouble width = DAY_VIEW_TIME_COLUMN_WIDTH;
+	gdouble day_width;
+	gdouble font_size, max_font_size, hour_font_size, minute_font_size;
+	gchar buf[20];
+	const gchar *minute;
+	const gint LONG_EVENT_OFFSET = 6;
+	gboolean use_24_hour;
+	gint i, hour, row;
+	gdouble hour_minute_xl, hour_minute_xr;
+	cairo_t *cr;
+
+	model = gnome_calendar_get_model (gcal);
+	use_24_hour = calendar_config_get_24_hour_format ();
+
+	/* Fill the left time column in light-gray. */
+	print_border (context, left, left + width, top, bottom, -1.0, 0.9);
+	/* Fill the right time column in light-gray */
+	print_border (context, right - width, right, top, bottom, -1.0, 0.9);
+
+	/* Draw the border around the entire view. */
+	cr = gtk_print_context_get_cairo_context (context);
+
+	cairo_set_source_rgb (cr, 0, 0, 0);
+	print_border (context, left, right, top, bottom, 1.0, -1.0);
+
+	/* Draw the vertical line on the right of the time column. */
+	cr = gtk_print_context_get_cairo_context (context);
+	cairo_set_line_width (cr, 0.0);
+	cairo_move_to (cr, left + width, bottom);
+	cairo_line_to (cr, left + width, top);
+	cairo_stroke (cr);
+
+	cairo_move_to (cr, right - width, bottom);
+	cairo_line_to (cr, right - width, top);
+	cairo_stroke (cr);
+
+	/* Calculate the row height. */
+	if (top > bottom)
+		yinc = (top - bottom) / (pdi->end_hour - pdi->start_hour);
+	else
+		yinc = (bottom - top) / (pdi->end_hour - pdi->start_hour);
+
+        /* Get the 2 fonts we need. */
+	font_size = yinc * 0.6;
+	max_font_size = width * 0.45;
+	hour_font_size = MIN (font_size, max_font_size);
+	font_hour = get_font_for_size (hour_font_size, PANGO_WEIGHT_BOLD);
+
+	font_size = yinc * 0.33;
+	max_font_size = width * 0.2;
+	minute_font_size = MIN (font_size, max_font_size);
+	font_minute = get_font_for_size (minute_font_size, PANGO_WEIGHT_BOLD);
+	hour_minute_xr = evo_calendar_print_renderer_get_width (
+		context, font_minute, use_24_hour ? "00" : _("am"));
+	if (!use_24_hour)
+		hour_minute_xr =
+			MAX (hour_minute_xr,
+			evo_calendar_print_renderer_get_width (
+			context, font_minute, _("pm")));
+
+	row = 0;
+	hour_minute_xl = left + width - hour_minute_xr - 3;
+	hour_minute_xr = right - hour_minute_xr - 3;
+	for (i = pdi->start_hour; i < pdi->end_hour; i++) {
+		y = top + yinc * (row + 1);
+		cr = gtk_print_context_get_cairo_context (context);
+		cairo_set_source_rgb (cr, 0, 0, 0);
+
+		if (use_24_hour) {
+			hour = i;
+			minute = "00";
+		} else {
+			if (i < 12)
+				minute = _("am");
+			else
+				minute = _("pm");
+
+			hour = i % 12;
+			if (hour == 0)
+				hour = 12;
+		}
+
+		/* the hour label/minute */
+		sprintf (buf, "%d", hour);
+		print_text (context, font_hour, buf, PANGO_ALIGN_RIGHT,
+			    left, hour_minute_xl,
+			    y - yinc, y - yinc + hour_font_size);
+		print_text (context, font_minute, minute, PANGO_ALIGN_LEFT,
+			    hour_minute_xl, left + width - 3,
+			    y - yinc, y - yinc + minute_font_size);
+
+		/* To the right */
+		print_text (context, font_hour, buf, PANGO_ALIGN_RIGHT,
+			    right - width, hour_minute_xr,
+			    y - yinc, y - yinc + hour_font_size);
+		print_text (context, font_minute, minute, PANGO_ALIGN_LEFT,
+			    hour_minute_xr, right - 3,
+			    y - yinc, y - yinc + minute_font_size);
+
+                /* Draw the horizontal line between hours, across the entire
+		   width of the day view. */
+		cr = gtk_print_context_get_cairo_context (context);
+		cairo_move_to (cr, left, y);
+		cairo_line_to (cr, right, y);
+		cairo_set_line_width (cr, 1);
+		cairo_stroke (cr);
+
+		/* Draw the horizontal line for the 1/2-hours, across the
+		   entire width except for part of the time column. */
+		cairo_move_to (cr, left + width * 0.6, y - yinc / 2);
+		cairo_line_to (cr, right, y - yinc / 2);
+		cairo_set_line_width (cr, 1);
+		cairo_stroke (cr);
+		row++;
+	}
+
+	/* Draw the vertical lines for the days */
+	day_width = (right - left - 2*width) / pdi->days_shown;
+	for (i = 0; i < pdi->days_shown - 1; ++i) {
+	  cr = gtk_print_context_get_cairo_context (context);
+	  cairo_move_to (cr, left + width + day_width * (i + 1), top);
+	  cairo_line_to (cr, left + width + day_width * (i + 1), bottom);
+	  cairo_set_line_width (cr, 1);
+	  cairo_stroke (cr);
+	}
+
+	/* And now the ones from the border to the hours, looks weird otherwise */
+	cr = gtk_print_context_get_cairo_context (context);
+	cairo_move_to (cr, left, HEADER_HEIGHT);
+	cairo_line_to (cr, left, HEADER_HEIGHT + DAY_VIEW_ROW_HEIGHT + LONG_EVENT_OFFSET);
+
+	cairo_move_to (cr, right, HEADER_HEIGHT);
+	cairo_line_to (cr, right, HEADER_HEIGHT + DAY_VIEW_ROW_HEIGHT + LONG_EVENT_OFFSET);
+	cairo_stroke (cr);
+
+	pango_font_description_free (font_hour);
+	pango_font_description_free (font_minute);
+}
+
+static void
+print_work_week_day_details (GtkPrintContext *context, GnomeCalendar *gcal,
+			     time_t whence, double left, double right,
+			     double top, double bottom, struct pdinfo *_pdi)
+{
+	ECalModel *model;
+	icaltimezone *zone;
+	EDayViewEvent *event;
+	PangoFontDescription *font;
+	time_t start, end;
+	struct pdinfo pdi = { 0 };
+	gint rows_in_top_display, i;
+	gdouble font_size, max_font_size;
+	cairo_t *cr;
+	GdkPixbuf *pixbuf = NULL;
+#define LONG_DAY_EVENTS_TOP_SPACING 4
+#define LONG_DAY_EVENTS_BOTTOM_SPACING 2
+
+	model = gnome_calendar_get_model (gcal);
+	zone = e_cal_model_get_timezone (model);
+
+	start = time_day_begin_with_zone (whence, zone);
+	end = time_day_end_with_zone (start, zone);
+
+	pdi.days_shown = 1;
+	pdi.day_starts[0] = start;
+	pdi.day_starts[1] = end;
+	pdi.long_events = g_array_new (FALSE, FALSE, sizeof (EDayViewEvent));
+	pdi.events[0] = g_array_new (FALSE, FALSE, sizeof (EDayViewEvent));
+	pdi.start_hour = calendar_config_get_day_start_hour ();
+	pdi.end_hour = calendar_config_get_day_end_hour ();
+	if (calendar_config_get_day_end_minute () != 0)
+		pdi.end_hour++;
+	pdi.rows = (pdi.end_hour - pdi.start_hour) * 2;
+	pdi.mins_per_row = 30;
+	pdi.start_minute_offset = pdi.start_hour * 60;
+	pdi.end_minute_offset = pdi.end_hour * 60;
+	pdi.use_24_hour_format = calendar_config_get_24_hour_format ();
+	pdi.zone = e_cal_model_get_timezone (model);
+
+	/* Get the events from the server. */
+	e_cal_model_generate_instances (model, start, end, print_day_details_cb, &pdi);
+	qsort (pdi.long_events->data, pdi.long_events->len,
+	       sizeof (EDayViewEvent), e_day_view_event_sort_func);
+	qsort (pdi.events[0]->data, pdi.events[0]->len,
+	       sizeof (EDayViewEvent), e_day_view_event_sort_func);
+
+	pdi.start_hour = MIN (pdi.start_hour, _pdi->start_hour);
+	pdi.end_hour = MAX (pdi.end_hour, _pdi->end_hour);
+
+	/* TODO: This should be redundant */
+	/* Also print events outside of work hours */
+	if (pdi.events[0]->len > 0) {
+		struct icaltimetype tt;
+
+		event = &g_array_index (pdi.events[0], EDayViewEvent, 0);
+		tt = icaltime_from_timet_with_zone (event->start, FALSE, zone);
+		if (tt.hour < pdi.start_hour)
+			pdi.start_hour = tt.hour;
+		pdi.start_minute_offset = pdi.start_hour * 60;
+
+		event = &g_array_index (pdi.events[0], EDayViewEvent, pdi.events[0]->len - 1);
+		tt = icaltime_from_timet_with_zone (event->end, FALSE, zone);
+		if (tt.hour > pdi.end_hour || tt.hour == 0) {
+			pdi.end_hour = tt.hour ? tt.hour : 24;
+			if (tt.minute > 0)
+				pdi.end_hour++;
+		}
+		pdi.end_minute_offset = pdi.end_hour * 60;
+
+		pdi.rows = (pdi.end_hour - pdi.start_hour) * 2;
+	}
+
+	/* Lay them out the long events, across the top of the page. */
+	e_day_view_layout_long_events (pdi.long_events, pdi.days_shown,
+				       pdi.day_starts, &rows_in_top_display);
+
+	 /*Print the long events. */
+	font = get_font_for_size (12, PANGO_WEIGHT_NORMAL);
+
+	/* We always leave space for DAY_VIEW_MIN_ROWS_IN_TOP_DISPLAY in the
+	   top display, but we may have more rows than that, in which case
+	   the main display area will be compressed. */
+	/* Limit long day event to half the height of the panel */
+	rows_in_top_display = MIN (MAX (rows_in_top_display,
+				   DAY_VIEW_MIN_ROWS_IN_TOP_DISPLAY),
+				   (bottom-top)*0.5/DAY_VIEW_ROW_HEIGHT);
+
+	if (rows_in_top_display > pdi.long_events->len)
+		rows_in_top_display = pdi.long_events->len;
+
+	for (i = 0; i < rows_in_top_display && i < pdi.long_events->len; i++) {
+		event = &g_array_index (pdi.long_events, EDayViewEvent, i);
+		print_day_long_event (
+			context, font, left, right,
+			top + LONG_DAY_EVENTS_TOP_SPACING, bottom,
+			DAY_VIEW_ROW_HEIGHT, event, &pdi, model);
+	}
+
+	if (rows_in_top_display < pdi.long_events->len) {
+		/* too many events */
+		cairo_t *cr = gtk_print_context_get_cairo_context (context);
+		gint x, y;
+
+		if (!pixbuf) {
+			const gchar **xpm = (const gchar **)jump_xpm;
+
+			/* this ugly thing is here only to get rid of compiler warning
+			   about unused 'jump_xpm_focused' */
+			if (pixbuf)
+				xpm = (const gchar **)jump_xpm_focused;
+
+			pixbuf = gdk_pixbuf_new_from_xpm_data (xpm);
+		}
+
+		/* Right align - 10 comes from print_day_long_event  too */
+		x = right - gdk_pixbuf_get_width (pixbuf) * 0.5 - 10;
+		/* Placing '...' over the last all day event entry printed. '-1 -1' comes
+			from print_long_day_event (top/bottom spacing in each cell) */
+		y = top + LONG_DAY_EVENTS_TOP_SPACING
+			+ DAY_VIEW_ROW_HEIGHT * (i - 1)
+			+ (DAY_VIEW_ROW_HEIGHT - 1 - 1) * 0.5;
+
+		cairo_save (cr);
+		cairo_scale (cr, 0.5, 0.5);
+		gdk_cairo_set_source_pixbuf (cr, pixbuf, x * 2.0, y * 2.0);
+		cairo_paint (cr);
+		cairo_restore (cr);
+	}
+
+	if (!rows_in_top_display)
+		rows_in_top_display++;
+
+	/* Draw the border around the long events. */
+	cr = gtk_print_context_get_cairo_context (context);
+
+	cairo_set_source_rgb (cr, 0, 0, 0);
+	print_border (
+		context, left, right,
+		top, top + rows_in_top_display * DAY_VIEW_ROW_HEIGHT +
+		LONG_DAY_EVENTS_TOP_SPACING + LONG_DAY_EVENTS_BOTTOM_SPACING,
+		1.0, -1.0);
+
+	/* Adjust the area containing the main display. */
+	top += rows_in_top_display * DAY_VIEW_ROW_HEIGHT
+		+ LONG_DAY_EVENTS_TOP_SPACING
+		+ LONG_DAY_EVENTS_BOTTOM_SPACING;
+
+	/* lay out the short events, within the day. */
+	e_day_view_layout_day_events (pdi.events[0], DAY_VIEW_ROWS,
+				      DAY_VIEW_MINS_PER_ROW, pdi.cols_per_row, -1);
+
+	/* print the short events. */
+	if (top > bottom )
+		max_font_size = ((top - bottom) / pdi.rows) - 4;
+	else
+		max_font_size = ((bottom - top ) / pdi.rows) - 4;
+	font_size = MIN (DAY_NORMAL_FONT_SIZE, max_font_size);
+	font = get_font_for_size (font_size, PANGO_WEIGHT_NORMAL);
+
+	for (i = 0; i < pdi.events[0]->len; i++) {
+		event = &g_array_index (pdi.events[0], EDayViewEvent, i);
+		print_day_event (context, font, left,
+				 right, top, bottom, event, &pdi, model);
+	}
+
+	/* Free everything. */
+	if (pixbuf)
+		g_object_unref (pixbuf);
+	free_event_array (pdi.long_events);
+	pango_font_description_free (font);
+	g_array_free (pdi.long_events, TRUE);
+	free_event_array (pdi.events[0]);
+	g_array_free (pdi.events[0], TRUE);
+}
+
+/* Figure out what the overal hour limits are */
+static gboolean
+print_work_week_view_cb (ECalComponent *comp,
+                         time_t istart,
+                         time_t iend,
+                         gpointer data)
+{
+	ECalModelGenerateInstancesData *mdata = (ECalModelGenerateInstancesData *) data;
+	struct pdinfo *pdi = (struct pdinfo *) mdata->cb_data;
+	struct icaltimetype tt;
+
+	tt = icaltime_from_timet_with_zone (istart, FALSE, pdi->zone);
+	pdi->start_hour = MIN (pdi->start_hour, tt.hour);
+
+	tt = icaltime_from_timet_with_zone (iend, FALSE, pdi->zone);
+	/* If we're past the hour, use the next one */
+	pdi->end_hour = MAX (pdi->end_hour, tt.minute ? tt.hour + 1 : tt.hour);
+
+	return TRUE;
+}
+
+static void
+print_work_week_view (GtkPrintContext *context, GnomeCalendar *gcal, time_t date)
+{
+	GtkPageSetup *setup;
+	icaltimezone *zone;
+	time_t when, start, end;
+	gdouble width, height, l;
+	gdouble small_month_width;
+	gdouble weeknum_inc;
+	gint i, days = 5;
+	gchar buf[100];
+	const gint LONG_EVENT_OFFSET = 6;
+	struct pdinfo pdi = { 0 };
+	struct tm tm;
+	gdouble day_width, day_x;
+	ECalModel *model;
+
+	model = gnome_calendar_get_model (gcal);
+	zone = e_cal_model_get_timezone (model);
+
+	setup = gtk_print_context_get_page_setup (context);
+
+	width = gtk_page_setup_get_page_width (setup, GTK_UNIT_POINTS);
+	height = gtk_page_setup_get_page_height (setup, GTK_UNIT_POINTS);
+
+	small_month_width = calc_small_month_width (context, HEADER_HEIGHT);
+	weeknum_inc = get_show_week_numbers () ? small_month_width / 7.0 : 0;
+
+	/* We always start on a Monday */
+	start = time_week_begin_with_zone (date, 1, zone);
+	end = time_add_day_with_zone (start, days, zone);
+
+	pdi.days_shown = days;
+	pdi.start_hour = calendar_config_get_day_start_hour ();
+	pdi.end_hour = calendar_config_get_day_end_hour ();
+	pdi.zone = zone;
+
+	e_cal_model_generate_instances (model, start, end,
+					print_work_week_view_cb, &pdi);
+
+	print_work_week_background (context, gcal, date, &pdi, 0.0, width,
+				    HEADER_HEIGHT + DAY_VIEW_ROW_HEIGHT + LONG_EVENT_OFFSET,
+				    height);
+
+	print_border (context, 0.0, width, 0.0, HEADER_HEIGHT, 1.0, 0.9);
+
+	/* Print the 2 mini calendar-months. */
+	l = width - SMALL_MONTH_PAD - (small_month_width + weeknum_inc) * 2 -
+		SMALL_MONTH_SPACING;
+
+	print_month_small (context, gcal, start,
+			   l, 4, l + small_month_width + weeknum_inc, HEADER_HEIGHT + 4,
+			   DATE_MONTH | DATE_YEAR, start, end, FALSE);
+
+	l += SMALL_MONTH_SPACING + small_month_width + weeknum_inc;
+	print_month_small (context, gcal,
+			   time_add_month_with_zone (start, 1, zone),
+			   l, 4, l + small_month_width + weeknum_inc, HEADER_HEIGHT + 4,
+			   DATE_MONTH | DATE_YEAR, 0, 0, FALSE);
+
+	/* Print the start day of the week, e.g. '7th May 2001'. */
+	convert_timet_to_struct_tm (start, zone, &tm);
+	format_date (&tm, DATE_DAY | DATE_MONTH | DATE_YEAR, buf, 100);
+	print_text_size_bold (context, buf, PANGO_ALIGN_LEFT,
+			      3, width,
+			      4, 4 + 24);
+
+	/* Print the end day of the week, e.g. '13th May 2001'. */
+	/* We need to substract one or the wrong day will be printed */
+	convert_timet_to_struct_tm (
+		time_add_day_with_zone (end, -1, zone), zone, &tm);
+	format_date (&tm, DATE_DAY | DATE_MONTH | DATE_YEAR, buf, 100);
+	print_text_size_bold (context, buf, PANGO_ALIGN_LEFT,
+			      3, width,
+			      24 + 3, 24 + 3 + 24);
+
+	/* Now print each days' events */
+	day_width = (width - 2*DAY_VIEW_TIME_COLUMN_WIDTH) / days;
+	when = start;
+	for (i = 0; i < days; ++i) {
+		day_x = DAY_VIEW_TIME_COLUMN_WIDTH + day_width * i;
+
+		/* Print the day, e.g. 'Tuesday'. */
+		convert_timet_to_struct_tm (when, zone, &tm);
+		format_date (&tm, DATE_DAYNAME, buf, 100);
+
+		print_text_size_bold (context, buf, PANGO_ALIGN_LEFT,
+				      day_x + 4, day_x + day_width,
+				      HEADER_HEIGHT + 4, HEADER_HEIGHT + 4 + 18);
+
+		print_work_week_day_details (context, gcal, when,
+					     day_x, day_x + day_width,
+					     HEADER_HEIGHT, height, &pdi);
+		when = time_add_day_with_zone (when, 1, zone);
+	}
+}
+
+static void
 print_week_view (GtkPrintContext *context, GnomeCalendar *gcal, time_t date)
 {
 	GtkPageSetup *setup;
-	icaltimezone *zone = calendar_config_get_icaltimezone ();
-	gdouble l, week_numbers_inc;
+	ECalModel *model;
+	icaltimezone *zone;
+	gdouble l, week_numbers_inc, small_month_width;
 	gchar buf[100];
 	time_t when;
 	gint week_start_day;
@@ -2260,10 +2758,14 @@ print_week_view (GtkPrintContext *context, GnomeCalendar *gcal, time_t date)
 
 	width = gtk_page_setup_get_page_width (setup, GTK_UNIT_POINTS);
 	height = gtk_page_setup_get_page_height (setup, GTK_UNIT_POINTS);
-	week_numbers_inc = get_show_week_numbers () ? SMALL_MONTH_WIDTH / 7.0 : 0;
+	small_month_width = calc_small_month_width (context, HEADER_HEIGHT);
+	week_numbers_inc = get_show_week_numbers () ? small_month_width / 7.0 : 0;
 
-	tm = *convert_timet_to_struct_tm (date, zone);
-	week_start_day = calendar_config_get_week_start_day ();
+	model = gnome_calendar_get_model (gcal);
+	zone = e_cal_model_get_timezone (model);
+
+	convert_timet_to_struct_tm (date, zone, &tm);
+	week_start_day = e_cal_model_get_week_start_day (model);
 	when = time_week_begin_with_zone (date, week_start_day, zone);
 
 	/* If the week starts on a Sunday, we have to show the Saturday first,
@@ -2287,32 +2789,34 @@ print_week_view (GtkPrintContext *context, GnomeCalendar *gcal, time_t date)
 
 	/* Print the border around the header area. */
 	print_border (context, 0.0, width,
-                      0.0, HEADER_HEIGHT + 2.0 + 20, 1.0, 0.9);
+		      0.0, HEADER_HEIGHT + 2.0 + 20, 1.0, 0.9);
 
 	/* Print the 2 mini calendar-months. */
-	l = width - SMALL_MONTH_PAD - (SMALL_MONTH_WIDTH + week_numbers_inc) * 2
+	l = width - SMALL_MONTH_PAD - (small_month_width + week_numbers_inc) * 2
 		- SMALL_MONTH_SPACING;
 	print_month_small (context, gcal, when,
-			   l, 4, l + SMALL_MONTH_WIDTH + week_numbers_inc, HEADER_HEIGHT + 10,
+			   l, 4, l + small_month_width + week_numbers_inc, HEADER_HEIGHT + 10,
 			   DATE_MONTH | DATE_YEAR, when,
 			   time_add_week_with_zone (when, 1, zone), FALSE);
 
-	l += SMALL_MONTH_SPACING + SMALL_MONTH_WIDTH + week_numbers_inc;
+	l += SMALL_MONTH_SPACING + small_month_width + week_numbers_inc;
 	print_month_small (context, gcal,
 			   time_add_month_with_zone (when, 1, zone),
-			   l, 4, l + SMALL_MONTH_WIDTH + week_numbers_inc, HEADER_HEIGHT + 10,
+			   l, 4, l + small_month_width + week_numbers_inc, HEADER_HEIGHT + 10,
 			   DATE_MONTH | DATE_YEAR, when,
 			   time_add_week_with_zone (when, 1, zone), FALSE);
 
 	/* Print the start day of the week, e.g. '7th May 2001'. */
-	format_date (when, DATE_DAY | DATE_MONTH | DATE_YEAR, buf, 100);
+	convert_timet_to_struct_tm (when, zone, &tm);
+	format_date (&tm, DATE_DAY | DATE_MONTH | DATE_YEAR, buf, 100);
 	print_text_size_bold (context, buf, PANGO_ALIGN_LEFT,
 			      3, width,
 			      4, 4 + 24);
 
 	/* Print the end day of the week, e.g. '13th May 2001'. */
 	when = time_add_day_with_zone (when, 6, zone);
-	format_date (when, DATE_DAY | DATE_MONTH | DATE_YEAR, buf, 100);
+	convert_timet_to_struct_tm (when, zone, &tm);
+	format_date (&tm, DATE_DAY | DATE_MONTH | DATE_YEAR, buf, 100);
 	print_text_size_bold (context, buf, PANGO_ALIGN_LEFT,
 			      3, width,
 			      24 + 3, 24 + 3 + 24);
@@ -2321,17 +2825,23 @@ print_week_view (GtkPrintContext *context, GnomeCalendar *gcal, time_t date)
 static void
 print_month_view (GtkPrintContext *context, GnomeCalendar *gcal, time_t date)
 {
+	ECalModel *model;
 	GtkPageSetup *setup;
-	icaltimezone *zone = calendar_config_get_icaltimezone ();
+	icaltimezone *zone;
 	gchar buf[100];
 	gdouble width, height;
-	gdouble l, week_numbers_inc;
+	gdouble l, week_numbers_inc, small_month_width;
+	struct tm tm;
+
+	model = gnome_calendar_get_model (gcal);
+	zone = e_cal_model_get_timezone (model);
 
 	setup = gtk_print_context_get_page_setup (context);
 
 	width = gtk_page_setup_get_page_width (setup, GTK_UNIT_POINTS);
 	height = gtk_page_setup_get_page_height (setup, GTK_UNIT_POINTS);
-	week_numbers_inc = get_show_week_numbers () ? SMALL_MONTH_WIDTH / 7.0 : 0;
+	small_month_width = calc_small_month_width (context, HEADER_HEIGHT);
+	week_numbers_inc = get_show_week_numbers () ? small_month_width / 7.0 : 0;
 
 	/* Print the main month view. */
 	print_month_summary (context, gcal, date, 0.0, width, HEADER_HEIGHT, height);
@@ -2339,65 +2849,34 @@ print_month_view (GtkPrintContext *context, GnomeCalendar *gcal, time_t date)
 	/* Print the border around the header. */
 	print_border (context, 0.0, width, 0.0, HEADER_HEIGHT + 10, 1.0, 0.9);
 
-	l = width - SMALL_MONTH_PAD - SMALL_MONTH_WIDTH - week_numbers_inc;
+	l = width - SMALL_MONTH_PAD - small_month_width - week_numbers_inc;
 
 	/* Print the 2 mini calendar-months. */
 	print_month_small (context, gcal,
 			   time_add_month_with_zone (date, 1, zone),
-			   l, 4, l + SMALL_MONTH_WIDTH + week_numbers_inc, HEADER_HEIGHT + 4,
+			   l, 4, l + small_month_width + week_numbers_inc, HEADER_HEIGHT + 4,
 			   DATE_MONTH | DATE_YEAR, 0, 0, FALSE);
 
 	print_month_small (context, gcal,
 			   time_add_month_with_zone (date, -1, zone),
-			   8, 4, width / 7 + 20, HEADER_HEIGHT + 4,
+			   8, 4, 8 + small_month_width + week_numbers_inc, HEADER_HEIGHT + 4,
 			   DATE_MONTH | DATE_YEAR, 0, 0, FALSE);
 
 	/* Print the month, e.g. 'May 2001'. */
-	format_date (date, DATE_MONTH | DATE_YEAR, buf, 100);
+	convert_timet_to_struct_tm (date, zone, &tm);
+	format_date (&tm, DATE_MONTH | DATE_YEAR, buf, 100);
 	print_text_size_bold (context, buf, PANGO_ALIGN_CENTER,
 			      3, width - 3,
-                              3, 3 + 24);
+			      3, 3 + 24);
 
 }
 
-/* XXX Evolution doesn't have a "year" view. */
-#if 0
-static void
-print_year_view (GtkPrintContext *context, GnomeCalendar *gcal, time_t date)
-{
-	GtkPageSetup *setup;
-	gchar buf[100];
-	cairo_t *cr;
-	gdouble width, height;
-
-	setup = gtk_print_context_get_page_setup (context);
-
-	width = gtk_page_setup_get_page_width (setup, GTK_UNIT_POINTS);
-	height = gtk_page_setup_get_page_height (setup, GTK_UNIT_POINTS);
-
-	cr = gtk_print_context_get_cairo_context (context);
-
-	cairo_show_page (cr);
-	print_year_summary (context, gcal, date, 0.0,
-			    width, 50,
-			    height, TRUE);
-
-	/* centered title */
-	format_date (date, DATE_YEAR, buf, 100);
-	print_text_size_bold (context, buf, PANGO_ALIGN_CENTER,
-			      3, width,
-			      3, 27);
-	cr=gtk_print_context_get_cairo_context (context);
-	cairo_show_page (cr);
-}
-#endif
-
 static gboolean
 same_date (struct tm tm1, time_t t2, icaltimezone *zone)
 {
 	struct tm tm2;
 
-	tm2 = *convert_timet_to_struct_tm (t2, zone);
+	convert_timet_to_struct_tm (t2, zone, &tm2);
 
 	return
 	    tm1.tm_mday == tm2.tm_mday &&
@@ -2407,32 +2886,31 @@ same_date (struct tm tm1, time_t t2, icaltimezone *zone)
 
 static void
 write_label_piece (time_t t,
-		   time_t *start_cmp,
+                   time_t *start_cmp,
+                   icaltimezone *zone,
+                   gboolean use_24_hour_format,
                    gchar *buffer,
                    gint size,
                    gchar *stext,
                    const gchar *etext)
 {
-	icaltimezone *zone = calendar_config_get_icaltimezone ();
 	struct tm tmp_tm;
 	gint len;
 
-	tmp_tm = *convert_timet_to_struct_tm (t, zone);
+	convert_timet_to_struct_tm (t, zone, &tmp_tm);
 
 	if (stext != NULL)
 		strcat (buffer, stext);
 
 	len = strlen (buffer);
 	if (start_cmp && same_date (tmp_tm, *start_cmp, zone))
-		e_time_format_time (&tmp_tm,
-				    calendar_config_get_24_hour_format (),
-				    FALSE,
-				    &buffer[len], size - len);
+		e_time_format_time (
+			&tmp_tm, use_24_hour_format,
+			FALSE, &buffer[len], size - len);
 	else
-		e_time_format_date_and_time (&tmp_tm,
-					     calendar_config_get_24_hour_format (),
-					     FALSE, FALSE,
-					     &buffer[len], size - len);
+		e_time_format_date_and_time (
+			&tmp_tm, use_24_hour_format, FALSE,
+			FALSE, &buffer[len], size - len);
 	if (etext != NULL)
 		strcat (buffer, etext);
 }
@@ -2456,8 +2934,15 @@ get_zone_from_tzid (ECal *client, const gchar *tzid)
 }
 
 static void
-print_date_label (GtkPrintContext *context, ECalComponent *comp, ECal *client,
-		  double left, double right, double top, double bottom)
+print_date_label (GtkPrintContext *context,
+                  ECalComponent *comp,
+                  ECal *client,
+                  icaltimezone *zone,
+                  gboolean use_24_hour_format,
+                  gdouble left,
+                  gdouble right,
+                  gdouble top,
+                  gdouble bottom)
 {
 	icaltimezone *start_zone, *end_zone, *due_zone, *completed_zone;
 	ECalComponentDateTime datetime;
@@ -2468,7 +2953,7 @@ print_date_label (GtkPrintContext *context, ECalComponent *comp, ECal *client,
 	if (datetime.value) {
 		start_zone = get_zone_from_tzid (client, datetime.tzid);
 		if (!start_zone || datetime.value->is_date)
-			start_zone = calendar_config_get_icaltimezone ();
+			start_zone = zone;
 		start = icaltime_as_timet_with_zone (*datetime.value,
 						     start_zone);
 	}
@@ -2478,7 +2963,7 @@ print_date_label (GtkPrintContext *context, ECalComponent *comp, ECal *client,
 	if (datetime.value) {
 		end_zone = get_zone_from_tzid (client, datetime.tzid);
 		if (!end_zone || datetime.value->is_date)
-			end_zone = calendar_config_get_icaltimezone ();
+			end_zone = zone;
 		end = icaltime_as_timet_with_zone (*datetime.value,
 						   end_zone);
 	}
@@ -2488,7 +2973,7 @@ print_date_label (GtkPrintContext *context, ECalComponent *comp, ECal *client,
 	if (datetime.value) {
 		due_zone = get_zone_from_tzid (client, datetime.tzid);
 		if (!due_zone || datetime.value->is_date)
-			due_zone = calendar_config_get_icaltimezone ();
+			due_zone = zone;
 		due = icaltime_as_timet_with_zone (*datetime.value,
 						   due_zone);
 	}
@@ -2505,12 +2990,16 @@ print_date_label (GtkPrintContext *context, ECalComponent *comp, ECal *client,
 	buffer[0] = '\0';
 
 	if (start > 0)
-		write_label_piece (start, NULL, buffer, 1024, NULL, NULL);
+		write_label_piece (
+			start, NULL, zone, use_24_hour_format,
+			buffer, 1024, NULL, NULL);
 
 	if (end > 0 && start > 0) {
 		/* Translators: This is part of "START to END" text,
 		 * where START and END are date/times. */
-		write_label_piece (end, &start, buffer, 1024, _(" to "), NULL);
+		write_label_piece (
+			end, &start, zone, use_24_hour_format,
+			buffer, 1024, _(" to "), NULL);
 	}
 
 	if (complete > 0) {
@@ -2518,11 +3007,15 @@ print_date_label (GtkPrintContext *context, ECalComponent *comp, ECal *client,
 			/* Translators: This is part of "START to END
 			 * (Completed COMPLETED)", where COMPLETED is a
 			 * completed date/time. */
-			write_label_piece (complete, NULL, buffer, 1024, _(" (Completed "), ")");
+			write_label_piece (
+				complete, NULL, zone, use_24_hour_format,
+				buffer, 1024, _(" (Completed "), ")");
 		} else {
 			/* Translators: This is part of "Completed COMPLETED",
 			 * where COMPLETED is a completed date/time. */
-			write_label_piece (complete, &start, buffer, 1024, _("Completed "), NULL);
+			write_label_piece (
+				complete, &start, zone, use_24_hour_format,
+				buffer, 1024, _("Completed "), NULL);
 		}
 	}
 
@@ -2530,12 +3023,16 @@ print_date_label (GtkPrintContext *context, ECalComponent *comp, ECal *client,
 		if (start > 0) {
 			/* Translators: This is part of "START (Due DUE)",
 			 * where START and DUE are dates/times. */
-			write_label_piece (due, NULL, buffer, 1024, _(" (Due "), ")");
+			write_label_piece (
+				due, NULL, zone, use_24_hour_format,
+				buffer, 1024, _(" (Due "), ")");
 		} else {
 			/* Translators: This is part of "Due DUE",
 			 * where DUE is a date/time due the event
 			 * should be finished. */
-			write_label_piece (due, &start, buffer, 1024, _("Due "), NULL);
+			write_label_piece (
+				due, &start, zone, use_24_hour_format,
+				buffer, 1024, _("Due "), NULL);
 		}
 	}
 
@@ -2554,6 +3051,8 @@ print_calendar_draw_page (GtkPrintOperation *operation,
 			print_day_view (context, pcali->gcal, pcali->start);
 			break;
 		case GNOME_CAL_WORK_WEEK_VIEW:
+			print_work_week_view (context, pcali->gcal, pcali->start);
+			break;
 		case GNOME_CAL_WEEK_VIEW:
 			print_week_view (context, pcali->gcal, pcali->start);
 			break;
@@ -2612,7 +3111,7 @@ print_calendar (GnomeCalendar *gcal,
 	operation = e_print_operation_new ();
 	gtk_print_operation_set_n_pages (operation, 1);
 
-        g_signal_connect (
+	g_signal_connect (
 		operation, "draw_page",
 		G_CALLBACK (print_calendar_draw_page), &pcali);
 
@@ -2716,7 +3215,10 @@ print_comp_draw_real (GtkPrintOperation *operation,
 
 	/* Date information */
 	if (page_nr == 0)
-		print_date_label (context, comp, client, 0.0, width, top + 3, top + 15);
+		print_date_label (
+			context, comp, client,
+			pci->zone, pci->use_24_hour_format,
+			0.0, width, top + 3, top + 15);
 	top += 20;
 
 	/* Attendees */
@@ -2902,7 +3404,9 @@ print_comp_begin_print (GtkPrintOperation *operation,
 }
 
 void
-print_comp (ECalComponent *comp, ECal *client, GtkPrintOperationAction action)
+print_comp (ECalComponent *comp,
+            ECal *client,
+            GtkPrintOperationAction action)
 {
 	GtkPrintOperation *operation;
 	PrintCompItem pci;
@@ -2911,6 +3415,8 @@ print_comp (ECalComponent *comp, ECal *client, GtkPrintOperationAction action)
 
 	pci.comp = comp;
 	pci.client = client;
+	pci.zone = calendar_config_get_icaltimezone ();
+	pci.use_24_hour_format = calendar_config_get_24_hour_format ();
 
 	operation = e_print_operation_new ();
 	gtk_print_operation_set_n_pages (operation, 1);
@@ -2949,7 +3455,8 @@ print_title (GtkPrintContext *context, const gchar *text, gdouble page_width)
 
 	cairo_move_to (cr, 0.0, 0.0);
 	pango_cairo_show_layout (cr, layout);
-
+	cairo_translate (cr, 0.0, 18);
+	cairo_save (cr);
 	cairo_restore (cr);
 
 	g_object_unref (layout);
