commit a1eb9cce5334833eb116297bc6eeb475301d5262
Author: Milan Crha <mcrha@redhat.com>
Date:   Mon Jan 10 08:36:02 2011 -0500

    evolution-2.28.3-mail-to-task-crash.patch

diff --git a/mail/em-folder-view.c b/mail/em-folder-view.c
index b792cd5..d4ee01d 100644
--- a/mail/em-folder-view.c
+++ b/mail/em-folder-view.c
@@ -2142,14 +2142,12 @@ emfv_enable_menus(EMFolderView *emfv)
 	if (emfv->uic == NULL)
 		return;
 
-	{
-		if (emfv->menu && emfv->folder) {
-			EMMenuTargetSelect *t;
+	if (emfv->menu) {
+		EMMenuTargetSelect *t;
 
-			t = em_menu_target_new_select(emfv->menu, emfv->folder, emfv->folder_uri, message_list_get_selected(emfv->list));
-			t->target.widget = GTK_WIDGET (emfv);
-			e_menu_update_target((EMenu *)emfv->menu, t);
-		}
+		t = em_menu_target_new_select(emfv->menu, emfv->folder, emfv->folder ? emfv->folder_uri : NULL, emfv->folder ? message_list_get_selected(emfv->list) : NULL);
+		t->target.widget = GTK_WIDGET (emfv);
+		e_menu_update_target((EMenu *)emfv->menu, t);
 	}
 
 	if (emfv->folder) {
diff --git a/plugins/mail-to-task/mail-to-task.c b/plugins/mail-to-task/mail-to-task.c
index 94b629e..9f04142 100644
--- a/plugins/mail-to-task/mail-to-task.c
+++ b/plugins/mail-to-task/mail-to-task.c
@@ -922,30 +922,40 @@ e_plugin_lib_enable (EPluginLib *ep, gint enable)
 void
 org_gnome_mail_to_event (gpointer ep, EMPopupTargetSelect *t)
 {
+	g_return_if_fail (t != NULL);
+
 	mail_to_event (E_CAL_SOURCE_TYPE_EVENT, FALSE, t->uids, t->folder, (EMFolderView *) t->target.widget);
 }
 
 void
 org_gnome_mail_to_event_menu (EPlugin *ep, EMMenuTargetSelect *t)
 {
+	g_return_if_fail (t != NULL);
+
 	mail_to_event (E_CAL_SOURCE_TYPE_EVENT, FALSE, t->uids, t->folder, (EMFolderView *) t->target.widget);
 }
 
 void
 org_gnome_mail_to_meeting (gpointer ep, EMPopupTargetSelect *t)
 {
+	g_return_if_fail (t != NULL);
+
 	mail_to_event (E_CAL_SOURCE_TYPE_EVENT, TRUE, t->uids, t->folder, (EMFolderView *) t->target.widget);
 }
 
 void
 org_gnome_mail_to_meeting_menu (EPlugin *ep, EMMenuTargetSelect *t)
 {
+	g_return_if_fail (t != NULL);
+
 	mail_to_event (E_CAL_SOURCE_TYPE_EVENT, TRUE, t->uids, t->folder, (EMFolderView *) t->target.widget);
 }
 
 void
 org_gnome_mail_to_task (gpointer ep, EMPopupTargetSelect *t)
 {
+	g_return_if_fail (t != NULL);
+
 	/* do not create assigned tasks */
 	mail_to_event (E_CAL_SOURCE_TYPE_TODO, FALSE, t->uids, t->folder, (EMFolderView *) t->target.widget);
 }
@@ -953,6 +963,8 @@ org_gnome_mail_to_task (gpointer ep, EMPopupTargetSelect *t)
 void
 org_gnome_mail_to_task_menu (EPlugin *ep, EMMenuTargetSelect *t)
 {
+	g_return_if_fail (t != NULL);
+
 	/* do not create assigned tasks */
 	mail_to_event (E_CAL_SOURCE_TYPE_TODO, FALSE, t->uids, t->folder, (EMFolderView *) t->target.widget);
 }
@@ -960,6 +972,8 @@ org_gnome_mail_to_task_menu (EPlugin *ep, EMMenuTargetSelect *t)
 void
 org_gnome_mail_to_memo (gpointer ep, EMPopupTargetSelect *t)
 {
+	g_return_if_fail (t != NULL);
+
 	/* do not set organizer and attendees for memos */
 	mail_to_event (E_CAL_SOURCE_TYPE_JOURNAL, FALSE, t->uids, t->folder, (EMFolderView *) t->target.widget);
 }
@@ -967,6 +981,8 @@ org_gnome_mail_to_memo (gpointer ep, EMPopupTargetSelect *t)
 void
 org_gnome_mail_to_memo_menu (EPlugin *ep, EMMenuTargetSelect *t)
 {
+	g_return_if_fail (t != NULL);
+
 	/* do not set organizer and attendees for memos */
 	mail_to_event (E_CAL_SOURCE_TYPE_JOURNAL, FALSE, t->uids, t->folder, (EMFolderView *) t->target.widget);
 }
