diff -up evolution-2.28.3/calendar/gui/print.c.implicit-declarations evolution-2.28.3/calendar/gui/print.c
--- evolution-2.28.3/calendar/gui/print.c.implicit-declarations	2013-01-09 14:23:15.992297526 +0100
+++ evolution-2.28.3/calendar/gui/print.c	2013-01-09 14:23:45.863692549 +0100
@@ -1919,7 +1919,7 @@ print_week_summary (GtkPrintContext *con
 
 	/* Get a few config settings. */
 	if (multi_week_view)
-		psi.compress_weekend = e_cal_model_get_compress_weekend (model);
+		psi.compress_weekend = calendar_config_get_compress_weekend ();
 	else
 		psi.compress_weekend = TRUE;
 	psi.use_24_hour_format = e_cal_model_get_use_24_hour_format (model);
@@ -2018,7 +2018,7 @@ print_month_summary (GtkPrintContext *co
 	model = gnome_calendar_get_calendar_model (gcal);
 	zone = e_cal_model_get_timezone (model);
 	weekday = calendar_config_get_week_start_day ();
-	compress_weekend = e_cal_model_get_compress_weekend (model);
+	compress_weekend = calendar_config_get_compress_weekend ();
 
 	date = 0;
 	weeks = 6;
diff -up evolution-2.28.3/composer/e-msg-composer.c.implicit-declarations evolution-2.28.3/composer/e-msg-composer.c
--- evolution-2.28.3/composer/e-msg-composer.c.implicit-declarations	2013-01-09 14:23:16.125299284 +0100
+++ evolution-2.28.3/composer/e-msg-composer.c	2013-01-09 14:23:56.151828615 +0100
@@ -3118,7 +3118,7 @@ file_is_blacklisted (const gchar *argume
 	parts = g_strsplit (filename, G_DIR_SEPARATOR_S, -1);
 	n_parts = g_strv_length (parts);
 
-	for (ii = 0; ii < G_N_ELEMENT (blacklist); ii++) {
+	for (ii = 0; ii < G_N_ELEMENTS (blacklist); ii++) {
 		for (jj = 0; jj < n_parts; jj++) {
 			if (g_str_has_prefix (parts[jj], blacklist[ii])) {
 				blacklisted = TRUE;
