diff -up evolution-2.32.3/addressbook/gui/widgets/e-addressbook-view.c.covscan-issues evolution-2.32.3/addressbook/gui/widgets/e-addressbook-view.c
--- evolution-2.32.3/addressbook/gui/widgets/e-addressbook-view.c.covscan-issues	2011-03-29 20:38:45.000000000 +0200
+++ evolution-2.32.3/addressbook/gui/widgets/e-addressbook-view.c	2013-06-12 15:09:16.880355749 +0200
@@ -1419,7 +1419,7 @@ e_addressbook_view_delete_selection(EAdd
 	}
 
 	/* Sets the cursor, at the row after the deleted row */
-	if (GAL_IS_VIEW_MINICARD (gal_view) && row != 0) {
+	if (GAL_IS_VIEW_MINICARD (gal_view) && row != 0 && selection_model) {
 		select = e_sorter_model_to_sorted (selection_model->sorter, row);
 
 	/* Sets the cursor, before the deleted row if its the last row */
diff -up evolution-2.32.3/addressbook/gui/widgets/ea-minicard.c.covscan-issues evolution-2.32.3/addressbook/gui/widgets/ea-minicard.c
--- evolution-2.32.3/addressbook/gui/widgets/ea-minicard.c.covscan-issues	2011-03-29 20:38:45.000000000 +0200
+++ evolution-2.32.3/addressbook/gui/widgets/ea-minicard.c	2013-06-12 15:09:16.880355749 +0200
@@ -159,7 +159,7 @@ ea_minicard_get_name (AtkObject *accessi
 
 	/* if there exist no enough space for remain info, return */
 	if (new_str->len >= BUFFERSIZE) {
-		strncpy (name, new_str->str, BUFFERSIZE);
+		strncpy (name, new_str->str, BUFFERSIZE - 1);
 		return name;
 	}
 
diff -up evolution-2.32.3/calendar/gui/e-day-view-time-item.c.covscan-issues evolution-2.32.3/calendar/gui/e-day-view-time-item.c
--- evolution-2.32.3/calendar/gui/e-day-view-time-item.c.covscan-issues	2011-03-29 20:38:46.000000000 +0200
+++ evolution-2.32.3/calendar/gui/e-day-view-time-item.c	2013-06-12 15:09:16.889355749 +0200
@@ -493,9 +493,7 @@ edvti_draw_zone (GnomeCanvasItem   *canv
 			cairo_restore (cr);
 
 			if (show_midnight_date) {
-				strcpy (buffer, midnight_day);
-				strcat (buffer, " ");
-				strcat (buffer, midnight_month);
+				g_snprintf (buffer, sizeof (buffer), "%s %s", midnight_day, midnight_month);
 			} else if (e_cal_model_get_use_24_hour_format (model)) {
 				g_snprintf (buffer, sizeof (buffer), "%i:%02i",
 					    display_hour, minute);
@@ -529,7 +527,7 @@ edvti_draw_zone (GnomeCanvasItem   *canv
 				cairo_save (cr);
 				gdk_cairo_set_source_color (cr, &dark);
 				if (show_midnight_date)
-					strcpy (buffer, midnight_day);
+					g_snprintf (buffer, sizeof (buffer), "%s", midnight_day);
 				else
 					g_snprintf (buffer, sizeof (buffer), "%i",
 						    display_hour);
@@ -574,7 +572,7 @@ edvti_draw_zone (GnomeCanvasItem   *canv
 				/* In 12-hour format we display 'am' or 'pm'
 				   instead of '00'. */
 				if (show_midnight_date)
-					strcpy (buffer, midnight_month);
+					g_snprintf (buffer, sizeof (buffer), "%s", midnight_month);
 				else if (minute == 0
 				    && !e_cal_model_get_use_24_hour_format (model)) {
 					strcpy (buffer, suffix);
diff -up evolution-2.32.3/calendar/gui/print.c.covscan-issues evolution-2.32.3/calendar/gui/print.c
--- evolution-2.32.3/calendar/gui/print.c.covscan-issues	2013-06-12 15:09:16.797355749 +0200
+++ evolution-2.32.3/calendar/gui/print.c	2013-06-12 15:09:16.889355749 +0200
@@ -565,32 +565,34 @@ format_date (struct tm *tm,
              gchar *buffer,
              gint bufflen)
 {
-	gchar fmt[64];
+	GString *fmt;
 
-	fmt[0] = 0;
+	fmt = g_string_new ("");
 	if (flags & DATE_DAYNAME) {
-		strcat(fmt, "%A");
+		g_string_append (fmt, "%A");
 	}
 	if (flags & DATE_DAY) {
 		if (flags & DATE_DAYNAME)
-			strcat(fmt, " ");
-		strcat (fmt, gettext (days[tm->tm_mday-1]));
+			g_string_append (fmt, " ");
+		g_string_append (fmt, gettext (days[tm->tm_mday-1]));
 	}
 	if (flags & DATE_MONTH) {
 		if (flags & (DATE_DAY|DATE_DAYNAME))
-			strcat(fmt, " ");
-		strcat(fmt, "%B");
+			g_string_append (fmt, " ");
+		g_string_append (fmt, "%B");
 		if ((flags & (DATE_DAY|DATE_YEAR)) == (DATE_DAY|DATE_YEAR))
-			strcat(fmt, ",");
+			g_string_append (fmt, ",");
 	}
 	if (flags & DATE_YEAR) {
 		if (flags & (DATE_DAY|DATE_DAYNAME|DATE_MONTH))
-			strcat(fmt, " ");
-		strcat(fmt, "%Y");
+			g_string_append (fmt, " ");
+		g_string_append (fmt, "%Y");
 	}
-	e_utf8_strftime (buffer, bufflen, fmt, tm);
+	e_utf8_strftime (buffer, bufflen, fmt->str, tm);
 	buffer[bufflen - 1] = '\0';
 
+	g_string_free (fmt, TRUE);
+
 	return buffer;
 }
 
diff -up evolution-2.32.3/capplet/settings/mail-account-view.c.covscan-issues evolution-2.32.3/capplet/settings/mail-account-view.c
--- evolution-2.32.3/capplet/settings/mail-account-view.c.covscan-issues	2011-03-29 20:38:46.000000000 +0200
+++ evolution-2.32.3/capplet/settings/mail-account-view.c	2013-06-12 15:09:16.887355749 +0200
@@ -938,10 +938,12 @@ emae_check_servers (const gchar *email)
 	/* FIXME: Find a way to free the provider once given to account settings. */
 	provider->email = (gchar *)email;
 	tmp = strchr(email, '@');
-	tmp++;
+	if (tmp)
+		tmp++;
 	provider->domain = tmp;
 	tmp = strchr(dupe, '@');
-	*tmp = 0;
+	if (tmp)
+		*tmp = 0;
 	provider->username = (gchar *)g_quark_to_string(g_quark_from_string(dupe));
 	g_free(dupe);
 
diff -up evolution-2.32.3/capplet/settings/mail-guess-servers.c.covscan-issues evolution-2.32.3/capplet/settings/mail-guess-servers.c
--- evolution-2.32.3/capplet/settings/mail-guess-servers.c.covscan-issues	2011-03-29 20:38:46.000000000 +0200
+++ evolution-2.32.3/capplet/settings/mail-guess-servers.c	2013-06-12 15:09:16.887355749 +0200
@@ -72,9 +72,11 @@ xml_to_gchar (xmlChar *xml, EmailProvide
 		else if (*tmp == 'D')
 			repl = provider->domain;
 		sec_part = strstr(tmp, "\%");
-		sec_part++;
-		if (!*sec_part)
-			sec_part = "";
+		if (sec_part) {
+			sec_part++;
+			if (!*sec_part)
+				sec_part = "";
+		}
 
 		gxml = g_strdup_printf("%s%s%s", gxml ? gxml : (gchar *)xml, repl, sec_part);
 		tmp = strstr (gxml, "\%EMAIL");
diff -up evolution-2.32.3/capplet/settings/mail-view.c.covscan-issues evolution-2.32.3/capplet/settings/mail-view.c
--- evolution-2.32.3/capplet/settings/mail-view.c.covscan-issues	2011-03-29 20:38:46.000000000 +0200
+++ evolution-2.32.3/capplet/settings/mail-view.c	2013-06-12 15:09:16.887355749 +0200
@@ -490,6 +490,9 @@ mail_view_add_page (MailView *mv, guint1
 		 child = mail_view_add_account (mv, data, block);
 		break;
 	}
+
+	g_return_val_if_fail (child != NULL, NULL);
+
 	gtk_widget_grab_focus((GtkWidget *)child);
 	child->type = type;
 	g_signal_handlers_unblock_by_func(mv, mv_switch, mv);
diff -up evolution-2.32.3/composer/e-msg-composer.c.covscan-issues evolution-2.32.3/composer/e-msg-composer.c
--- evolution-2.32.3/composer/e-msg-composer.c.covscan-issues	2013-06-12 15:09:16.818355749 +0200
+++ evolution-2.32.3/composer/e-msg-composer.c	2013-06-12 15:09:16.886355749 +0200
@@ -1066,7 +1066,7 @@ build_message (EMsgComposer *composer,
 					TRUE, account->smime_encrypt_key);
 			}
 
-			camel_cipher_sign (cipher, account->smime_sign_key,
+			camel_cipher_sign (cipher, account ? account->smime_sign_key : NULL,
 				account_hash_algo_to_camel_hash (
 					(account != NULL) ?
 					e_account_get_string (account,
diff -up evolution-2.32.3/filter/e-filter-code.c.covscan-issues evolution-2.32.3/filter/e-filter-code.c
--- evolution-2.32.3/filter/e-filter-code.c.covscan-issues	2011-03-29 20:38:47.000000000 +0200
+++ evolution-2.32.3/filter/e-filter-code.c	2013-06-12 15:09:16.885355749 +0200
@@ -40,7 +40,11 @@ filter_code_build_code (EFilterElement *
 {
 	GList *l;
 	EFilterInput *fi = (EFilterInput *)element;
-	gboolean is_rawcode = fi && fi->type && g_str_equal (fi->type, "rawcode");
+	gboolean is_rawcode;
+
+	g_return_if_fail (fi != NULL);
+
+	is_rawcode = fi->type && g_str_equal (fi->type, "rawcode");
 
 	if (!is_rawcode)
 		g_string_append(out, "(match-all ");
diff -up evolution-2.32.3/libart_lgpl/art_render.c.covscan-issues evolution-2.32.3/libart_lgpl/art_render.c
--- evolution-2.32.3/libart_lgpl/art_render.c.covscan-issues	2011-03-29 20:38:48.000000000 +0200
+++ evolution-2.32.3/libart_lgpl/art_render.c	2013-06-12 15:09:16.884355749 +0200
@@ -150,8 +150,8 @@ art_render_composite (ArtRenderCallback 
   art_u32 run_alpha;
   art_u32 alpha;
   gint image_ix;
-  art_u16 src[ART_MAX_CHAN + 1];
-  art_u16 dst[ART_MAX_CHAN + 1];
+  art_u16 src[ART_MAX_CHAN + 1] = { 0, };
+  art_u16 dst[ART_MAX_CHAN + 1] = { 0, };
   gint n_chan = render->n_chan;
   ArtAlphaType alpha_type = render->alpha_type;
   gint n_ch = n_chan + (alpha_type != ART_ALPHA_NONE);
diff -up evolution-2.32.3/libart_lgpl/art_svp_vpath.c.covscan-issues evolution-2.32.3/libart_lgpl/art_svp_vpath.c
--- evolution-2.32.3/libart_lgpl/art_svp_vpath.c.covscan-issues	2011-03-29 20:38:48.000000000 +0200
+++ evolution-2.32.3/libart_lgpl/art_svp_vpath.c	2013-06-12 15:09:16.883355749 +0200
@@ -132,7 +132,7 @@ art_svp_from_vpath (ArtVpath *vpath)
       {
 	new_dir = (vpath[i].y > y ||
 		   (vpath[i].y == y && vpath[i].x > x)) ? 1 : -1;
-	if (dir && dir != new_dir)
+	if (dir && dir != new_dir && points)
 	  {
 	    /* new segment */
 	    x = points[n_points - 1].x;
diff -up evolution-2.32.3/libgnomecanvas/gnome-canvas-shape.c.covscan-issues evolution-2.32.3/libgnomecanvas/gnome-canvas-shape.c
--- evolution-2.32.3/libgnomecanvas/gnome-canvas-shape.c.covscan-issues	2011-03-29 20:38:48.000000000 +0200
+++ evolution-2.32.3/libgnomecanvas/gnome-canvas-shape.c	2013-06-12 15:09:16.888355749 +0200
@@ -979,7 +979,7 @@ gnome_canvas_shape_update_gdk (GnomeCanv
 			for (len = 0; vpath[len].code != ART_END; len++);
 
 			gnome_canvas_shape_ensure_gdk_points (gdk, len);
-			for (i = 0; i < len; i++) {
+			for (i = 0; i < len && gdk->points; i++) {
 				gdk->points[pos + i].x = (gint) floor (vpath[i].x + 0.5);
 				gdk->points[pos + i].y = (gint) floor (vpath[i].y + 0.5);
 
@@ -1021,7 +1021,7 @@ gnome_canvas_shape_update_gdk (GnomeCanv
 			for (len = 0; vpath[len].code != ART_END; len++);
 
 			gnome_canvas_shape_ensure_gdk_points (gdk, len);
-			for (i = 0; i < len; i++) {
+			for (i = 0; i < len && gdk->points; i++) {
 				gdk->points[pos + i].x = (gint) floor (vpath[i].x + 0.5);
 				gdk->points[pos + i].y = (gint) floor (vpath[i].y + 0.5);
 
diff -up evolution-2.32.3/libgnomecanvas/gnome-canvas-util.c.covscan-issues evolution-2.32.3/libgnomecanvas/gnome-canvas-util.c
--- evolution-2.32.3/libgnomecanvas/gnome-canvas-util.c.covscan-issues	2011-03-29 20:38:48.000000000 +0200
+++ evolution-2.32.3/libgnomecanvas/gnome-canvas-util.c	2013-06-12 15:09:16.888355749 +0200
@@ -281,7 +281,7 @@ gnome_canvas_polygon_to_point (gdouble *
 	best = 1.0e36;
 	intersections = 0;
 
-	for (i = num_points, p = poly; i > 1; i--, p += 2) {
+	for (i = num_points, p = poly; i > 1 && p; i--, p += 2) {
 		gdouble px, py, dist;
 
 		/* Compute the point on the current edge closest to the
diff -up evolution-2.32.3/mail/em-account-editor.c.covscan-issues evolution-2.32.3/mail/em-account-editor.c
--- evolution-2.32.3/mail/em-account-editor.c.covscan-issues	2013-06-12 15:09:16.791355749 +0200
+++ evolution-2.32.3/mail/em-account-editor.c	2013-06-12 15:09:16.885355749 +0200
@@ -3335,6 +3335,9 @@ emae_check_servers (const gchar *email)
 	gint len = G_N_ELEMENTS(mail_servers), i;
 	gchar *server = strchr (email, '@');
 
+	if (!server)
+		return NULL;
+
 	server++;
 
 	for (i=0; i<len; i++) {
@@ -3406,17 +3409,19 @@ emae_check_complete (EConfig *ec, const 
 		} else if (!strcmp (pageid, "10.receive")) {
 			if (!emae->priv->receive_set) {
 				ServerData *sdata;
-				gchar *user, *at;
+				gchar *user = NULL, *at;
 				gchar *uri = g_strdup (e_account_get_string (account, E_ACCOUNT_SOURCE_URL));
 				CamelURL *url;
 
 				emae->priv->receive_set = 1;
 				tmp = (gchar *)e_account_get_string (account, E_ACCOUNT_ID_ADDRESS);
 				at = strchr (tmp, '@');
-				user = g_alloca (at-tmp+1);
-				memcpy (user, tmp, at-tmp);
-				user[at-tmp] = 0;
-				at++;
+				if (at) {
+					user = g_alloca (at-tmp+1);
+					memcpy (user, tmp, at-tmp);
+					user[at-tmp] = 0;
+					at++;
+				}
 
 				sdata = emae->priv->selected_server = emae->emae_check_servers (tmp);
 				gtk_entry_set_text (emae->priv->source.username, sdata && sdata->recv_user && *sdata->recv_user ? sdata->recv_user : user);
@@ -3455,17 +3460,19 @@ emae_check_complete (EConfig *ec, const 
 		} else if (!strcmp (pageid, "30.send")) {
 			if (!emae->priv->send_set) {
 				CamelURL *url;
-				gchar *at, *user;
+				gchar *at, *user = NULL;
 				gchar *uri = (gchar *)e_account_get_string (account, E_ACCOUNT_TRANSPORT_URL);
 				ServerData *sdata;
 
 				emae->priv->send_set = 1;
 				tmp = e_account_get_string (account, E_ACCOUNT_ID_ADDRESS);
 				at = strchr (tmp, '@');
-				user = g_alloca (at-tmp+1);
-				memcpy (user, tmp, at-tmp);
-				user[at-tmp] = 0;
-				at++;
+				if (at) {
+					user = g_alloca (at-tmp+1);
+					memcpy (user, tmp, at-tmp);
+					user[at-tmp] = 0;
+					at++;
+				}
 
 				sdata = emae->priv->selected_server;
 				if (sdata != NULL && uri && (url = camel_url_new (uri, NULL)) != NULL) {
diff -up evolution-2.32.3/plugins/caldav/caldav-browse-server.c.covscan-issues evolution-2.32.3/plugins/caldav/caldav-browse-server.c
--- evolution-2.32.3/plugins/caldav/caldav-browse-server.c.covscan-issues	2011-03-29 20:38:48.000000000 +0200
+++ evolution-2.32.3/plugins/caldav/caldav-browse-server.c	2013-06-12 15:09:16.880355749 +0200
@@ -1323,13 +1323,15 @@ prepare_url (const gchar *server_url, gb
 		soup_uri_free (suri);
 	}
 
-	/* remove trailing slashes... */
-	len = strlen (url);
-	while (len--) {
-		if (url[len] == '/') {
-			url[len] = '\0';
-		} else {
-			break;
+	if (url) {
+		/* remove trailing slashes... */
+		len = strlen (url);
+		while (len--) {
+			if (url[len] == '/') {
+				url[len] = '\0';
+			} else {
+				break;
+			}
 		}
 	}
 
diff -up evolution-2.32.3/shell/e-shell-utils.c.covscan-issues evolution-2.32.3/shell/e-shell-utils.c
--- evolution-2.32.3/shell/e-shell-utils.c.covscan-issues	2013-06-12 15:09:16.838355749 +0200
+++ evolution-2.32.3/shell/e-shell-utils.c	2013-06-12 15:09:16.884355749 +0200
@@ -256,7 +256,7 @@ e_shell_run_save_dialog (EShell *shell,
 		gchar **flts = g_strsplit (filters, ";", -1);
 		gint i;
 
-		for (i = 0; flts[i]; i++) {
+		for (i = 0; flts && flts[i]; i++) {
 			GtkFileFilter *filter = gtk_file_filter_new ();
 			gchar *flt = flts[i];
 			gchar *delim = strchr (flt, ':'), *next = NULL;
diff -up evolution-2.32.3/shell/e-shell-view.c.covscan-issues evolution-2.32.3/shell/e-shell-view.c
--- evolution-2.32.3/shell/e-shell-view.c.covscan-issues	2011-03-29 20:38:57.000000000 +0200
+++ evolution-2.32.3/shell/e-shell-view.c	2013-06-12 15:09:16.884355749 +0200
@@ -451,6 +451,7 @@ shell_view_get_property (GObject *object
 			g_value_set_object (
 				value, e_shell_view_get_shell_backend (
 				E_SHELL_VIEW (object)));
+			return;
 
 		case PROP_SHELL_CONTENT:
 			g_value_set_object (
diff -up evolution-2.32.3/smclient/eggsmclient-xsmp.c.covscan-issues evolution-2.32.3/smclient/eggsmclient-xsmp.c
--- evolution-2.32.3/smclient/eggsmclient-xsmp.c.covscan-issues	2011-03-29 20:38:57.000000000 +0200
+++ evolution-2.32.3/smclient/eggsmclient-xsmp.c	2013-06-12 15:09:16.889355749 +0200
@@ -1165,6 +1165,7 @@ array_prop (const gchar *name, ...)
       pv.value = value;
       g_array_append_val (vals, pv);
     }
+  va_end (ap);
 
   prop->num_vals = vals->len;
   prop->vals = (SmPropValue *)vals->data;
diff -up evolution-2.32.3/widgets/misc/e-canvas.c.covscan-issues evolution-2.32.3/widgets/misc/e-canvas.c
--- evolution-2.32.3/widgets/misc/e-canvas.c.covscan-issues	2011-03-29 20:38:57.000000000 +0200
+++ evolution-2.32.3/widgets/misc/e-canvas.c	2013-06-12 15:09:16.883355749 +0200
@@ -657,7 +657,7 @@ e_canvas_item_grab_focus (GnomeCanvasIte
 		ev.focus_change.send_event = FALSE;
 		ev.focus_change.in = FALSE;
 
-		canvas_emit_event (item->canvas, &ev);
+		canvas_emit_event (focused_item->canvas, &ev);
 	}
 
 	item->canvas->focused_item = item;
diff -up evolution-2.32.3/widgets/table/e-tree.c.covscan-issues evolution-2.32.3/widgets/table/e-tree.c
--- evolution-2.32.3/widgets/table/e-tree.c.covscan-issues	2013-06-12 15:09:16.857355749 +0200
+++ evolution-2.32.3/widgets/table/e-tree.c	2013-06-12 15:09:16.882355749 +0200
@@ -2223,6 +2223,9 @@ e_tree_load_expanded_state_xml (ETree *e
 void
 e_tree_force_expanded_state (ETree *et, gint state)
 {
+	g_return_if_fail (et != NULL);
+	g_return_if_fail (E_IS_TREE (et));
+
 	e_tree_table_adapter_force_expanded_state (et->priv->etta, state);
 }
 
diff -up evolution-2.32.3/widgets/table/gal-a11y-e-cell.c.covscan-issues evolution-2.32.3/widgets/table/gal-a11y-e-cell.c
--- evolution-2.32.3/widgets/table/gal-a11y-e-cell.c.covscan-issues	2011-03-29 20:38:58.000000000 +0200
+++ evolution-2.32.3/widgets/table/gal-a11y-e-cell.c	2013-06-12 15:09:16.882355749 +0200
@@ -355,6 +355,7 @@ gal_a11y_e_cell_remove_action_by_name (G
 	}
 
 	g_return_val_if_fail (action_found, FALSE);
+	g_return_val_if_fail (list_node != NULL, FALSE);
 	_gal_a11y_e_cell_destroy_action_info (list_node->data, NULL);
 	cell->action_list = g_list_remove_link (cell->action_list, list_node);
 
diff -up evolution-2.32.3/widgets/table/gal-a11y-e-cell-tree.c.covscan-issues evolution-2.32.3/widgets/table/gal-a11y-e-cell-tree.c
--- evolution-2.32.3/widgets/table/gal-a11y-e-cell-tree.c.covscan-issues	2011-03-29 20:38:58.000000000 +0200
+++ evolution-2.32.3/widgets/table/gal-a11y-e-cell-tree.c	2013-06-12 15:09:16.881355749 +0200
@@ -92,7 +92,8 @@ ectr_subcell_weak_ref (GalA11yECellTree 
         if (subcell_a11y && subcell_view && subcell_view->kill_view_cb_data)
             subcell_view->kill_view_cb_data = g_list_remove (subcell_view->kill_view_cb_data, subcell_a11y);
 
-	g_signal_handler_disconnect (GAL_A11Y_E_CELL (a11y)->item->table_model,
+	if (a11y->model_row_changed_id)
+		g_signal_handler_disconnect (GAL_A11Y_E_CELL (a11y)->item->table_model,
 				     a11y->model_row_changed_id);
 	g_object_unref (a11y);
 }
@@ -244,10 +245,13 @@ gal_a11y_e_cell_tree_new (ETableItem *it
 				   model_col,
 				   view_col,
 				   row);
-	a11y->model_row_changed_id =
-		g_signal_connect (item->table_model, "model_row_changed",
+	if (subcell_a11y)
+		a11y->model_row_changed_id =
+			g_signal_connect (item->table_model, "model_row_changed",
 				  G_CALLBACK (ectr_model_row_changed_cb),
 				  subcell_a11y);
+	else
+		a11y->model_row_changed_id = 0;
 
         if (subcell_a11y && subcell_view)
         {
diff -up evolution-2.32.3/widgets/table/gal-a11y-e-table.c.covscan-issues evolution-2.32.3/widgets/table/gal-a11y-e-table.c
--- evolution-2.32.3/widgets/table/gal-a11y-e-table.c.covscan-issues	2011-03-29 20:38:58.000000000 +0200
+++ evolution-2.32.3/widgets/table/gal-a11y-e-table.c	2013-06-12 15:09:16.883355749 +0200
@@ -131,7 +131,7 @@ et_get_n_children (AtkObject *accessible
 
 	et = E_TABLE (gtk_accessible_get_widget (GTK_ACCESSIBLE (a11y)));
 
-	if (et->group) {
+	if (et && et->group) {
 		if (E_IS_TABLE_GROUP_LEAF (et->group))
 			n = 1;
 		else if (E_IS_TABLE_GROUP_CONTAINER (et->group)) {
diff -up evolution-2.32.3/widgets/table/gal-a11y-e-table-item.c.covscan-issues evolution-2.32.3/widgets/table/gal-a11y-e-table-item.c
--- evolution-2.32.3/widgets/table/gal-a11y-e-table-item.c.covscan-issues	2011-03-29 20:38:58.000000000 +0200
+++ evolution-2.32.3/widgets/table/gal-a11y-e-table-item.c	2013-06-12 15:09:16.882355749 +0200
@@ -1045,6 +1045,7 @@ gal_a11y_e_table_item_new (ETableItem *i
 	AtkObject *parent;
 	const gchar *name;
 
+	g_return_val_if_fail (item != NULL, NULL);
 	g_return_val_if_fail (item && item->cols >= 0 && item->rows >= 0, NULL);
 	a11y = g_object_new (gal_a11y_e_table_item_get_type (), NULL);
 
diff -up evolution-2.32.3/widgets/text/e-reflow.c.covscan-issues evolution-2.32.3/widgets/text/e-reflow.c
--- evolution-2.32.3/widgets/text/e-reflow.c.covscan-issues	2011-03-29 20:38:58.000000000 +0200
+++ evolution-2.32.3/widgets/text/e-reflow.c	2013-06-12 15:09:16.881355749 +0200
@@ -345,7 +345,7 @@ reflow_columns (EReflow *reflow)
 	reflow->columns = g_renew (int, reflow->columns, column_count);
 	column_count--;
 
-	for (; column_count > column_start; column_count--) {
+	for (; column_count > column_start && list; column_count--) {
 		GSList *to_free;
 		reflow->columns[column_count] = GPOINTER_TO_INT (list->data);
 		to_free = list;
