diff -up evolution-2.28.3/addressbook/gui/widgets/eab-contact-display.c.contact-qp-enc-print evolution-2.28.3/addressbook/gui/widgets/eab-contact-display.c
--- evolution-2.28.3/addressbook/gui/widgets/eab-contact-display.c.contact-qp-enc-print	2013-01-22 14:02:35.142073682 +0100
+++ evolution-2.28.3/addressbook/gui/widgets/eab-contact-display.c	2013-01-22 14:03:56.895071762 +0100
@@ -32,6 +32,8 @@
 #include "e-util/e-html-utils.h"
 #include "e-util/e-icon-factory.h"
 
+#include "addressbook/util/eab-book-util.h"
+
 #include <string.h>
 #include <glib/gi18n.h>
 #include <gtkhtml/gtkhtml.h>
diff -up evolution-2.28.3/addressbook/gui/widgets/eab-gui-util.c.contact-qp-enc-print evolution-2.28.3/addressbook/gui/widgets/eab-gui-util.c
--- evolution-2.28.3/addressbook/gui/widgets/eab-gui-util.c.contact-qp-enc-print	2013-01-22 14:02:54.848196004 +0100
+++ evolution-2.28.3/addressbook/gui/widgets/eab-gui-util.c	2013-01-22 14:03:56.895071762 +0100
@@ -43,9 +43,6 @@
 #include <composer/e-msg-composer.h>
 #include <mail/em-composer-utils.h>
 
-/* we link to camel for decoding quoted printable email addresses */
-#include <camel/camel-mime-utils.h>
-
 #include "addressbook/gui/contact-editor/eab-editor.h"
 #include "addressbook/gui/contact-editor/e-contact-editor.h"
 #include "addressbook/gui/contact-list-editor/e-contact-list-editor.h"
@@ -1031,57 +1028,3 @@ eab_create_image_chooser_widget(gchar *n
 
 	return w;
 }
-
-/* To parse something like...
-=?UTF-8?Q?=E0=A4=95=E0=A4=95=E0=A4=AC=E0=A5=82=E0=A5=8B=E0=A5=87?=\t\n=?UTF-8?Q?=E0=A4=B0?=\t\n<aa@aa.ccom>
-and return the decoded representation of name & email parts.
-*/
-gboolean
-eab_parse_qp_email (const gchar *string, gchar **name, gchar **email)
-{
-	struct _camel_header_address *address;
-	gboolean res = FALSE;
-
-	address = camel_header_address_decode (string, "UTF-8");
-
-	if (!address)
-		return FALSE;
-
-	/* report success only when we have filled both name and email address */
-	if (address->type == CAMEL_HEADER_ADDRESS_NAME  && address->name && *address->name && address->v.addr && *address->v.addr) {
-		*name = g_strdup (address->name);
-		*email = g_strdup (address->v.addr);
-		res = TRUE;
-	}
-
-	camel_header_address_unref (address);
-
-	return res;
-}
-
-/* This is only wrapper to parse_qp_mail, it decodes string and if returned TRUE,
-   then makes one string and returns it, otherwise returns NULL.
-   Returned string is usable to place directly into GtkHtml stream.
-   Returned value should be freed with g_free. */
-gchar *
-eab_parse_qp_email_to_html (const gchar *string)
-{
-	gchar *name = NULL, *mail = NULL;
-	gchar *html_name, *html_mail;
-	gchar *value;
-
-	if (!eab_parse_qp_email (string, &name, &mail))
-		return NULL;
-
-	html_name = e_text_to_html (name, 0);
-	html_mail = e_text_to_html (mail, E_TEXT_TO_HTML_CONVERT_ADDRESSES);
-
-	value = g_strdup_printf ("%s &lt;%s&gt;", html_name, html_mail);
-
-	g_free (html_name);
-	g_free (html_mail);
-	g_free (name);
-	g_free (mail);
-
-	return value;
-}
diff -up evolution-2.28.3/addressbook/gui/widgets/eab-gui-util.h.contact-qp-enc-print evolution-2.28.3/addressbook/gui/widgets/eab-gui-util.h
--- evolution-2.28.3/addressbook/gui/widgets/eab-gui-util.h.contact-qp-enc-print	2013-01-22 14:03:08.977074340 +0100
+++ evolution-2.28.3/addressbook/gui/widgets/eab-gui-util.h	2013-01-22 14:03:56.896071603 +0100
@@ -80,10 +80,6 @@ GtkWidget *eab_create_image_chooser_widg
 ESource            *eab_select_source             (const gchar *title, const gchar *message,
 						   const gchar *select_uid, GtkWindow *parent);
 
-/* To parse quoted printable address & return email & name fields */
-gboolean eab_parse_qp_email (const gchar *string, gchar **name, gchar **email);
-gchar *eab_parse_qp_email_to_html (const gchar *string);
-
 G_END_DECLS
 
 #endif /* __E_ADDRESSBOOK_UTIL_H__ */
diff -up evolution-2.28.3/addressbook/gui/widgets/e-addressbook-table-adapter.c.contact-qp-enc-print evolution-2.28.3/addressbook/gui/widgets/e-addressbook-table-adapter.c
--- evolution-2.28.3/addressbook/gui/widgets/e-addressbook-table-adapter.c.contact-qp-enc-print	2013-01-22 14:02:45.127073488 +0100
+++ evolution-2.28.3/addressbook/gui/widgets/e-addressbook-table-adapter.c	2013-01-22 14:03:56.895071762 +0100
@@ -24,6 +24,7 @@
 #include <glib/gi18n.h>
 #include "e-addressbook-model.h"
 #include "e-addressbook-table-adapter.h"
+#include "addressbook/util/eab-book-util.h"
 #include "eab-contact-merging.h"
 #include "eab-gui-util.h"
 #include <libxml/tree.h>
diff -up evolution-2.28.3/addressbook/gui/widgets/e-minicard.c.contact-qp-enc-print evolution-2.28.3/addressbook/gui/widgets/e-minicard.c
--- evolution-2.28.3/addressbook/gui/widgets/e-minicard.c.contact-qp-enc-print	2013-01-22 14:02:23.989195872 +0100
+++ evolution-2.28.3/addressbook/gui/widgets/e-minicard.c	2013-01-22 14:03:56.894074031 +0100
@@ -31,6 +31,7 @@
 #include <misc/e-canvas-utils.h>
 #include <misc/e-canvas.h>
 #include <libebook/e-book.h>
+#include "addressbook/util/eab-book-util.h"
 #include "eab-gui-util.h"
 #include "e-minicard.h"
 #include "e-minicard-label.h"
diff -up evolution-2.28.3/addressbook/printing/e-contact-print.c.contact-qp-enc-print evolution-2.28.3/addressbook/printing/e-contact-print.c
--- evolution-2.28.3/addressbook/printing/e-contact-print.c.contact-qp-enc-print	2013-01-22 14:03:20.210195941 +0100
+++ evolution-2.28.3/addressbook/printing/e-contact-print.c	2013-01-22 14:03:56.896071603 +0100
@@ -40,6 +40,8 @@
 #include "e-util/e-util.h"
 #include "e-util/e-util-private.h"
 
+#include "addressbook/util/eab-book-util.h"
+
 #include "e-contact-print.h"
 
 typedef struct _EContactPrintContext EContactPrintContext;
@@ -235,6 +237,41 @@ e_contact_start_new_column (EContactPrin
 	}
 }
 
+static gchar *
+get_contact_string_value (EContact *contact,
+			  gint field)
+{
+	const gchar *value;
+
+	g_return_val_if_fail (contact != NULL, NULL);
+
+	value = e_contact_get_const (contact, field);
+	if (!value || !*value)
+		return NULL;
+
+	if (field == E_CONTACT_EMAIL_1 ||
+	    field == E_CONTACT_EMAIL_2 ||
+	    field == E_CONTACT_EMAIL_3 ||
+	    field == E_CONTACT_EMAIL_4) {
+		gchar *email = NULL, *name = NULL;
+
+		if (eab_parse_qp_email (value, &name, &email)) {
+			gchar *res;
+
+			if (name && *name)
+				res = g_strdup_printf ("%s <%s>", name, email);
+			else
+				res = g_strdup_printf ("%s", email);
+
+			g_free (name);
+			g_free (email);
+
+			return res;
+		}
+	}
+	return g_strdup (value);
+}
+
 static gdouble
 e_contact_get_contact_height (EContact *contact, EContactPrintContext *ctxt)
 {
@@ -260,12 +297,14 @@ e_contact_get_contact_height (EContact *
 
 	for (field = E_CONTACT_FILE_AS; field != E_CONTACT_LAST_SIMPLE_STRING; field++)
 	{
-		const gchar *value;
+		gchar *value;
 		gchar *text;
 
-		value = e_contact_get_const (contact, field);
-		if (value == NULL || *value == '\0')
+		value = get_contact_string_value (contact, field);
+		if (value == NULL || *value == '\0') {
+			g_free (value);
 			continue;
+		}
 
 		text = g_strdup_printf ("%s:  %s",
 			e_contact_pretty_name (field), value);
@@ -275,6 +314,7 @@ e_contact_get_contact_height (EContact *
 
 		cntct_height += .2 * get_font_height (ctxt->style->body_font);
 
+		g_free (value);
 		g_free (text);
 	}
 
@@ -325,16 +365,18 @@ e_contact_print_contact (EContact *conta
 
 	for (field = E_CONTACT_FILE_AS; field != E_CONTACT_LAST_SIMPLE_STRING; field++)
 	{
-		const gchar *value;
+		gchar *value;
 		gchar *text;
 		gint wrapped_lines=0;
 
 		if (ctxt->y > page_height)
 			e_contact_start_new_column (ctxt);
 
-		value = e_contact_get_const (contact, field);
-		if (value == NULL || *value == '\0')
+		value = get_contact_string_value (contact, field);
+		if (value == NULL || *value == '\0') {
+			g_free (value);
 			continue;
+		}
 
 		text = g_strdup_printf ("%s:  %s",
 			e_contact_pretty_name (field), value);
@@ -351,6 +393,7 @@ e_contact_print_contact (EContact *conta
 
 		ctxt->y += .2 * get_font_height (ctxt->style->body_font);
 
+		g_free (value);
 		g_free (text);
 	}
 
diff -up evolution-2.28.3/addressbook/util/eab-book-util.c.contact-qp-enc-print evolution-2.28.3/addressbook/util/eab-book-util.c
--- evolution-2.28.3/addressbook/util/eab-book-util.c.contact-qp-enc-print	2013-01-22 14:03:47.156196729 +0100
+++ evolution-2.28.3/addressbook/util/eab-book-util.c	2013-01-22 14:03:56.897071425 +0100
@@ -28,6 +28,11 @@
 #include <glib.h>
 #include <glib-object.h>
 
+/* we link to camel for decoding quoted printable email addresses */
+#include <camel/camel-mime-utils.h>
+
+#include "e-util/e-html-utils.h"
+
 /*
  *
  * Specialized Queries
@@ -413,3 +418,57 @@ e_utf8_casefold_collate (const gchar *st
 {
 	return e_utf8_casefold_collate_len (str1, str2, -1);
 }
+
+/* To parse something like...
+=?UTF-8?Q?=E0=A4=95=E0=A4=95=E0=A4=AC=E0=A5=82=E0=A5=8B=E0=A5=87?=\t\n=?UTF-8?Q?=E0=A4=B0?=\t\n<aa@aa.ccom>
+and return the decoded representation of name & email parts.
+*/
+gboolean
+eab_parse_qp_email (const gchar *string, gchar **name, gchar **email)
+{
+	struct _camel_header_address *address;
+	gboolean res = FALSE;
+
+	address = camel_header_address_decode (string, "UTF-8");
+
+	if (!address)
+		return FALSE;
+
+	/* report success only when we have filled both name and email address */
+	if (address->type == CAMEL_HEADER_ADDRESS_NAME  && address->name && *address->name && address->v.addr && *address->v.addr) {
+		*name = g_strdup (address->name);
+		*email = g_strdup (address->v.addr);
+		res = TRUE;
+	}
+
+	camel_header_address_unref (address);
+
+	return res;
+}
+
+/* This is only wrapper to parse_qp_mail, it decodes string and if returned TRUE,
+   then makes one string and returns it, otherwise returns NULL.
+   Returned string is usable to place directly into GtkHtml stream.
+   Returned value should be freed with g_free. */
+gchar *
+eab_parse_qp_email_to_html (const gchar *string)
+{
+	gchar *name = NULL, *mail = NULL;
+	gchar *html_name, *html_mail;
+	gchar *value;
+
+	if (!eab_parse_qp_email (string, &name, &mail))
+		return NULL;
+
+	html_name = e_text_to_html (name, 0);
+	html_mail = e_text_to_html (mail, E_TEXT_TO_HTML_CONVERT_ADDRESSES);
+
+	value = g_strdup_printf ("%s &lt;%s&gt;", html_name, html_mail);
+
+	g_free (html_name);
+	g_free (html_mail);
+	g_free (name);
+	g_free (mail);
+
+	return value;
+}
diff -up evolution-2.28.3/addressbook/util/eab-book-util.h.contact-qp-enc-print evolution-2.28.3/addressbook/util/eab-book-util.h
--- evolution-2.28.3/addressbook/util/eab-book-util.h.contact-qp-enc-print	2013-01-22 14:03:36.748073847 +0100
+++ evolution-2.28.3/addressbook/util/eab-book-util.h	2013-01-22 14:03:56.896071603 +0100
@@ -58,6 +58,10 @@ void                   eab_query_address
 gint                    e_utf8_casefold_collate_len (const gchar *str1, const gchar *str2, gint len);
 gint                    e_utf8_casefold_collate (const gchar *str1, const gchar *str2);
 
+/* To parse quoted printable address & return email & name fields */
+gboolean eab_parse_qp_email (const gchar *string, gchar **name, gchar **email);
+gchar *eab_parse_qp_email_to_html (const gchar *string);
+
 G_END_DECLS
 
 #endif /* __EAB_UTIL_H__ */
