diff -up evolution-2.32.3/mail/message-list.c.wide-view-icons evolution-2.32.3/mail/message-list.c
--- evolution-2.32.3/mail/message-list.c.wide-view-icons	2013-08-12 17:07:31.179413100 +0200
+++ evolution-2.32.3/mail/message-list.c	2013-08-12 17:07:31.752413100 +0200
@@ -1864,7 +1864,7 @@ static ECell * create_composite_cell (gi
 	cell_hbox = e_cell_hbox_new ();
 
 	/* Exclude the meeting icon. */
-	cell_attach = e_cell_toggle_new (attachment_icons, 2);
+	cell_attach = e_cell_toggle_new (attachment_icons, G_N_ELEMENTS (attachment_icons));
 
 	cell_date = e_cell_date_new (NULL, GTK_JUSTIFY_RIGHT);
 	e_cell_date_set_format_component (E_CELL_DATE (cell_date), "mail");
diff -up evolution-2.32.3/widgets/table/e-tree.c.wide-view-icons evolution-2.32.3/widgets/table/e-tree.c
--- evolution-2.32.3/widgets/table/e-tree.c.wide-view-icons	2013-08-12 17:07:31.284413100 +0200
+++ evolution-2.32.3/widgets/table/e-tree.c	2013-08-12 17:07:31.749413100 +0200
@@ -3674,6 +3674,9 @@ e_tree_set_info_message (ETree *tree, co
 	gtk_widget_get_allocation (widget, &allocation);
 
 	if (!tree->priv->info_text) {
+		if (allocation.width < 60)
+			allocation.width = 60;
+
 		tree->priv->info_text = gnome_canvas_item_new (
 			GNOME_CANVAS_GROUP (gnome_canvas_root (tree->priv->table_canvas)),
 			e_text_get_type (),
