diff -up evolution-3.22.6/calendar/alarm-notify/notify-main.c.gtype-init-workaround evolution-3.22.6/calendar/alarm-notify/notify-main.c
--- evolution-3.22.6/calendar/alarm-notify/notify-main.c.gtype-init-workaround	2017-11-14 11:55:40.525214399 +0100
+++ evolution-3.22.6/calendar/alarm-notify/notify-main.c	2017-11-14 11:56:58.471213321 +0100
@@ -68,6 +68,11 @@ main (gint argc,
 	bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
 	textdomain (GETTEXT_PACKAGE);
 
+	/* Workaround https://bugzilla.gnome.org/show_bug.cgi?id=674885 */
+	g_type_ensure (G_TYPE_DBUS_CONNECTION);
+	g_type_ensure (G_TYPE_DBUS_PROXY);
+	g_type_ensure (G_BUS_TYPE_SESSION);
+
 	gtk_init (&argc, &argv);
 
 	alarm_notify_service = alarm_notify_new (NULL, &error);
diff -up evolution-3.22.6/shell/main.c.gtype-init-workaround evolution-3.22.6/shell/main.c
--- evolution-3.22.6/shell/main.c.gtype-init-workaround	2017-11-14 11:55:40.525214399 +0100
+++ evolution-3.22.6/shell/main.c	2017-11-14 11:57:34.104212829 +0100
@@ -459,6 +459,11 @@ main (gint argc,
 	/* Initialize timezone specific global variables */
 	tzset ();
 
+	/* Workaround https://bugzilla.gnome.org/show_bug.cgi?id=674885 */
+	g_type_ensure (G_TYPE_DBUS_CONNECTION);
+	g_type_ensure (G_TYPE_DBUS_PROXY);
+	g_type_ensure (G_BUS_TYPE_SESSION);
+
 	/* The contact maps feature uses clutter-gtk. */
 #ifdef WITH_CONTACT_MAPS
 	success = gtk_clutter_init_with_args (
