diff -up evolution-3.12.11/composer/e-msg-composer.c.misc-after-rebase-changes evolution-3.12.11/composer/e-msg-composer.c
--- evolution-3.12.11/composer/e-msg-composer.c.misc-after-rebase-changes	2014-10-29 12:11:43.000000000 +0100
+++ evolution-3.12.11/composer/e-msg-composer.c	2015-05-22 12:15:25.666735476 +0200
@@ -4542,7 +4542,7 @@ e_msg_composer_add_inline_image_from_fil
 	wrapper = camel_data_wrapper_new ();
 	camel_data_wrapper_construct_from_stream_sync (
 		wrapper, stream, NULL, NULL);
-	g_object_unref (CAMEL_OBJECT (stream));
+	g_object_unref (stream);
 
 	mime_type = e_util_guess_mime_type (dec_file_name, TRUE);
 	if (mime_type == NULL)
diff -up evolution-3.12.11/shell/main.c.misc-after-rebase-changes evolution-3.12.11/shell/main.c
--- evolution-3.12.11/shell/main.c.misc-after-rebase-changes	2015-01-06 15:05:43.000000000 +0100
+++ evolution-3.12.11/shell/main.c	2015-05-22 12:15:25.666735476 +0200
@@ -691,7 +691,7 @@ exit:
 	g_object_unref (shell);
 
 	if (e_shell_get_default () != NULL) {
-		g_warning ("Shell not finalized on exit");
+		/* g_warning ("Shell not finalized on exit"); */
 
 		/* To not run in the safe mode the next start */
 		if (e_file_lock_get_pid () == getpid ())
