diff -up evolution-3.12.11/addressbook/gui/widgets/eab-contact-formatter.c.coverity-scan evolution-3.12.11/addressbook/gui/widgets/eab-contact-formatter.c
--- evolution-3.12.11/addressbook/gui/widgets/eab-contact-formatter.c.coverity-scan	2014-07-17 12:47:57.000000000 +0200
+++ evolution-3.12.11/addressbook/gui/widgets/eab-contact-formatter.c	2015-05-22 10:25:53.896508227 +0200
@@ -401,7 +401,7 @@ render_title_block (EABContactFormatter
 		const gchar *uri = photo->data.uri;
 		/* WebKit 2.2.x doesn't re-escape URIs, thus do this for versions before and after this */
 		#if !(WEBKIT_MAJOR_VERSION == 2 && WEBKIT_MINOR_VERSION == 2)
-		gchar *unescaped = g_uri_unescape_string (photo->data.uri, NULL);
+		gchar *unescaped = g_uri_unescape_string (uri, NULL);
 		uri = unescaped;
 		#endif
 		g_string_append_printf (
@@ -882,7 +882,7 @@ render_compact (EABContactFormatter *for
 			const gchar *uri = photo->data.uri;
 			/* WebKit 2.2.x doesn't re-escape URIs, thus do this for versions before and after this */
 			#if !(WEBKIT_MAJOR_VERSION == 2 && WEBKIT_MINOR_VERSION == 2)
-			gchar *unescaped = g_uri_unescape_string (photo->data.uri, NULL);
+			gchar *unescaped = g_uri_unescape_string (uri, NULL);
 			uri = unescaped;
 			#endif
 			g_string_append_printf (
diff -up evolution-3.12.11/calendar/gui/e-calendar-view.c.coverity-scan evolution-3.12.11/calendar/gui/e-calendar-view.c
--- evolution-3.12.11/calendar/gui/e-calendar-view.c.coverity-scan	2014-12-09 16:01:59.000000000 +0100
+++ evolution-3.12.11/calendar/gui/e-calendar-view.c	2015-05-22 10:25:53.897508219 +0200
@@ -2023,7 +2023,7 @@ e_calendar_view_get_tooltips (const ECal
 {
 	GtkWidget *label, *box, *hbox, *ebox, *frame;
 	const gchar *str;
-	gchar *tmp, *tmp1, *tmp2;
+	gchar *tmp, *tmp1 = NULL, *tmp2 = NULL;
 	ECalComponentOrganizer organiser;
 	ECalComponentDateTime dtstart, dtend;
 	icalcomponent *clone_comp;
@@ -2181,6 +2181,8 @@ e_calendar_view_get_tooltips (const ECal
 			tmp = tmp2;
 			tmp2 = NULL;
 		}
+	} else {
+		tmp = NULL;
 	}
 
 	e_cal_component_free_datetime (&dtstart);
diff -up evolution-3.12.11/calendar/gui/e-day-view.c.coverity-scan evolution-3.12.11/calendar/gui/e-day-view.c
--- evolution-3.12.11/calendar/gui/e-day-view.c.coverity-scan	2014-12-22 16:05:29.000000000 +0100
+++ evolution-3.12.11/calendar/gui/e-day-view.c	2015-05-22 10:25:53.898508211 +0200
@@ -603,6 +603,8 @@ e_day_view_get_text_color (EDayView *day
 		color.blue = 65535.0f;
 	}
 
+	color.pixel = 0;
+
 	return color;
 }
 
diff -up evolution-3.12.11/e-util/e-cell-date-edit.c.coverity-scan evolution-3.12.11/e-util/e-cell-date-edit.c
--- evolution-3.12.11/e-util/e-cell-date-edit.c.coverity-scan	2014-04-24 19:00:26.000000000 +0200
+++ evolution-3.12.11/e-util/e-cell-date-edit.c	2015-05-22 10:25:53.898508211 +0200
@@ -960,8 +960,8 @@ e_cell_date_edit_update_cell (ECellDateE
 			ecell_text, ecv->e_table_model,
 			ecol->spec->model_col, ecp->popup_row, text);
 		e_cell_leave_edit (
-			ecv, ecp->popup_view_col,
-			ecol->spec->model_col, ecp->popup_row, NULL);
+			ecv, ecol->spec->model_col,
+			ecp->popup_view_col, ecp->popup_row, NULL);
 	}
 
 	e_cell_text_free_text (ecell_text, ecv->e_table_model,
diff -up evolution-3.12.11/e-util/e-filter-input.c.coverity-scan evolution-3.12.11/e-util/e-filter-input.c
--- evolution-3.12.11/e-util/e-filter-input.c.coverity-scan	2014-03-24 10:25:23.000000000 +0100
+++ evolution-3.12.11/e-util/e-filter-input.c	2015-05-22 10:25:53.898508211 +0200
@@ -107,7 +107,7 @@ filter_input_validate (EFilterElement *e
 
 		regfree (&regexpat);
 	} else if (!input->allow_empty && (!input->values || !input->values->next)) {
-		const gchar *value = input->values->data;
+		const gchar *value = input->values ? input->values->data : NULL;
 		gboolean is_empty = value == NULL;
 
 		if (value) {
diff -up evolution-3.12.11/e-util/e-focus-tracker.c.coverity-scan evolution-3.12.11/e-util/e-focus-tracker.c
--- evolution-3.12.11/e-util/e-focus-tracker.c.coverity-scan	2014-03-24 10:25:23.000000000 +0100
+++ evolution-3.12.11/e-util/e-focus-tracker.c	2015-05-22 10:25:53.899508203 +0200
@@ -1133,9 +1133,9 @@ e_focus_tracker_paste_clipboard (EFocusT
 
 	} else if (GTK_IS_TEXT_VIEW (focus)) {
 		GtkClipboard *clipboard;
-		GtkTextView *text_view = GTK_TEXT_VIEW (focus);
-		GtkTextBuffer *buffer = gtk_text_view_get_buffer (text_view);
-		gboolean is_editable = gtk_text_view_get_editable (text_view);
+		GtkTextView *text_view;
+		GtkTextBuffer *buffer;
+		gboolean is_editable;
 
 		clipboard = gtk_widget_get_clipboard (
 			focus, GDK_SELECTION_CLIPBOARD);
diff -up evolution-3.12.11/e-util/e-table-sort-info.c.coverity-scan evolution-3.12.11/e-util/e-table-sort-info.c
--- evolution-3.12.11/e-util/e-table-sort-info.c.coverity-scan	2014-03-24 10:25:23.000000000 +0100
+++ evolution-3.12.11/e-util/e-table-sort-info.c	2015-05-22 10:25:53.899508203 +0200
@@ -854,7 +854,7 @@ e_table_sort_info_save_to_node (ETableSo
 
 	for (ii = 0; ii < group_count; ii++) {
 		ETableColumnSpecification *column_spec;
-		GtkSortType sort_type;
+		GtkSortType sort_type = GTK_SORT_ASCENDING;
 		xmlNode *new_node;
 		gint index;
 
@@ -881,7 +881,7 @@ e_table_sort_info_save_to_node (ETableSo
 
 	for (ii = 0; ii < sort_count; ii++) {
 		ETableColumnSpecification *column_spec;
-		GtkSortType sort_type;
+		GtkSortType sort_type = GTK_SORT_ASCENDING;
 		xmlNode *new_node;
 		gint index;
 
diff -up evolution-3.12.11/mail/e-mail-reader-utils.c.coverity-scan evolution-3.12.11/mail/e-mail-reader-utils.c
--- evolution-3.12.11/mail/e-mail-reader-utils.c.coverity-scan	2014-12-12 13:46:11.000000000 +0100
+++ evolution-3.12.11/mail/e-mail-reader-utils.c	2015-05-22 10:25:53.899508203 +0200
@@ -970,6 +970,7 @@ mail_reader_print_parse_message_cb (GObj
 	if (local_error) {
 		g_warn_if_fail (g_error_matches (local_error, G_IO_ERROR, G_IO_ERROR_CANCELLED));
 
+		/* coverity[check_return] */
 		e_activity_handle_cancellation (activity, local_error);
 		g_clear_error (&local_error);
 		async_context_free (async_context);
diff -up evolution-3.12.11/mail/em-folder-tree.c.coverity-scan evolution-3.12.11/mail/em-folder-tree.c
--- evolution-3.12.11/mail/em-folder-tree.c.coverity-scan	2014-06-17 10:10:00.000000000 +0200
+++ evolution-3.12.11/mail/em-folder-tree.c	2015-05-22 10:25:53.900508195 +0200
@@ -1950,7 +1950,7 @@ tree_drag_data_get (GtkWidget *widget,
 	GtkTreeModel *model;
 	GtkTreePath *src_path;
 	CamelFolder *folder;
-	CamelStore *store;
+	CamelStore *store = NULL;
 	GtkTreeIter iter;
 	gchar *folder_name = NULL;
 	gchar *folder_uri;
diff -up evolution-3.12.11/mail/em-folder-tree-model.c.coverity-scan evolution-3.12.11/mail/em-folder-tree-model.c
--- evolution-3.12.11/mail/em-folder-tree-model.c.coverity-scan	2014-07-17 12:48:16.000000000 +0200
+++ evolution-3.12.11/mail/em-folder-tree-model.c	2015-05-22 10:25:53.900508195 +0200
@@ -1480,7 +1480,7 @@ folder_tree_model_update_status_icon (St
 	GtkTreePath *path;
 	GtkTreeIter iter;
 	GIcon *icon = NULL;
-	const gchar *icon_name;
+	const gchar *icon_name = NULL;
 	gboolean was_connecting;
 	gboolean host_reachable;
 
diff -up evolution-3.12.11/mail/message-list.c.coverity-scan evolution-3.12.11/mail/message-list.c
--- evolution-3.12.11/mail/message-list.c.coverity-scan	2015-02-04 08:07:15.000000000 +0100
+++ evolution-3.12.11/mail/message-list.c	2015-05-22 10:25:53.901508186 +0200
@@ -3027,7 +3027,7 @@ message_list_duplicate_value (ETreeModel
 
 				return res;
 			} else
-				return value;
+				return NULL;
 
 		default:
 			g_return_val_if_reached (NULL);
diff -up evolution-3.12.11/mail/test-mail-autoconfig.c.coverity-scan evolution-3.12.11/mail/test-mail-autoconfig.c
--- evolution-3.12.11/mail/test-mail-autoconfig.c.coverity-scan	2014-03-24 10:25:23.000000000 +0100
+++ evolution-3.12.11/mail/test-mail-autoconfig.c	2015-05-22 10:25:53.901508186 +0200
@@ -39,6 +39,8 @@ main (gint argc,
 		autoconfig = e_mail_autoconfig_new_sync (
 			registry, argv[1], NULL, &error);
 		g_object_unref (registry);
+	} else {
+		autoconfig = NULL;
 	}
 
 	/* Sanity check. */
