diff -up evolution-2.32.3/mail/em-account-editor.c.account-option-mnemonic evolution-2.32.3/mail/em-account-editor.c
--- evolution-2.32.3/mail/em-account-editor.c.account-option-mnemonic	2011-03-29 20:38:48.000000000 +0200
+++ evolution-2.32.3/mail/em-account-editor.c	2013-05-29 09:31:21.450613687 +0200
@@ -2427,8 +2427,10 @@ emae_option_checkspin (EMAccountEditorSe
 	check = g_object_new (gtk_check_button_get_type (), "label", pre, "use_underline", TRUE, "active", on, NULL);
 
 	spin = gtk_spin_button_new ((GtkAdjustment *)gtk_adjustment_new (def, min, max, 1, 1, 0), 1, 0);
-	if (post)
+	if (post) {
 		label = gtk_label_new_with_mnemonic (post);
+		gtk_label_set_mnemonic_widget (GTK_LABEL (label), check);
+	}
 	gtk_box_pack_start ((GtkBox *)hbox, check, FALSE, TRUE, 0);
 	gtk_box_pack_start ((GtkBox *)hbox, spin, FALSE, TRUE, 0);
 	if (label)
