diff -up evolution-3.12.11/mail/e-mail-request.c.svg-images-in-mail evolution-3.12.11/mail/e-mail-request.c
--- evolution-3.12.11/mail/e-mail-request.c.svg-images-in-mail	2014-03-24 10:25:23.000000000 +0100
+++ evolution-3.12.11/mail/e-mail-request.c	2015-05-05 17:38:42.797518974 +0200
@@ -156,6 +156,11 @@ handle_mail_request (GSimpleAsyncResult
 			dw = camel_medium_get_content (CAMEL_MEDIUM (mime_part));
 			g_return_if_fail (dw);
 
+			if (!mime_type) {
+				g_free (request->priv->mime_type);
+				request->priv->mime_type = camel_data_wrapper_get_mime_type (dw);
+			}
+
 			camel_data_wrapper_decode_to_output_stream_sync (
 				dw, output_stream, cancellable, NULL);
 
