diff -up evolution-3.12.11/mail/mail-send-recv.c.empty-send-receive evolution-3.12.11/mail/mail-send-recv.c
--- evolution-3.12.11/mail/mail-send-recv.c.empty-send-receive	2014-07-10 11:13:27.000000000 +0200
+++ evolution-3.12.11/mail/mail-send-recv.c	2016-03-23 16:34:45.919254645 +0100
@@ -1284,6 +1284,12 @@ send_receive (GtkWindow *parent,
 
 	g_list_free (scan);
 
+	if (g_hash_table_size (data->active) == 0) {
+		if (data->gd)
+			gtk_widget_destroy ((GtkWidget *) data->gd);
+		free_send_data ();
+	}
+
 	return send_recv_dialog;
 }
 
